<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Misc | Computational Statistics</title>
  <meta name="description" content="Class notes for Math 154 at Pomona College: Computational Statistics. The notes are based extensively on An Introduction to Statistical Learning by James, Witten, Hastie, and Tibshirani as well as Modern Data Science with R by Baumer, Kaplan, and Horton." />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Misc | Computational Statistics" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Class notes for Math 154 at Pomona College: Computational Statistics. The notes are based extensively on An Introduction to Statistical Learning by James, Witten, Hastie, and Tibshirani as well as Modern Data Science with R by Baumer, Kaplan, and Horton." />
  <meta name="github-repo" content="hardin47/website/Math154/" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Misc | Computational Statistics" />
  
  <meta name="twitter:description" content="Class notes for Math 154 at Pomona College: Computational Statistics. The notes are based extensively on An Introduction to Statistical Learning by James, Witten, Hastie, and Tibshirani as well as Modern Data Science with R by Baumer, Kaplan, and Horton." />
  

<meta name="author" content="Jo Hardin" />


<meta name="date" content="2020-07-21" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="unsup.html"/>
<link rel="next" href="compstat.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Statistics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Class Information</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#Sep3"><i class="fa fa-check"></i><b>1.1</b> 9/3/19 Agenda</a></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#course-logistics"><i class="fa fa-check"></i><b>1.2</b> Course Logistics</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#course-content"><i class="fa fa-check"></i><b>1.3</b> Course Content</a><ul>
<li class="chapter" data-level="1.3.1" data-path="intro.html"><a href="intro.html#topics"><i class="fa fa-check"></i><b>1.3.1</b> Topics</a></li>
<li class="chapter" data-level="1.3.2" data-path="intro.html"><a href="intro.html#vocabulary"><i class="fa fa-check"></i><b>1.3.2</b> Vocabulary</a></li>
<li class="chapter" data-level="1.3.3" data-path="intro.html"><a href="intro.html#the-workflow"><i class="fa fa-check"></i><b>1.3.3</b> The Workflow</a></li>
<li class="chapter" data-level="1.3.4" data-path="intro.html"><a href="intro.html#principles-for-the-data-science-process-tldr"><i class="fa fa-check"></i><b>1.3.4</b> Principles for the Data Science Process tl;dr</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="intro.html"><a href="intro.html#Sep5"><i class="fa fa-check"></i><b>1.4</b> 9/5/19 Agenda</a></li>
<li class="chapter" data-level="1.5" data-path="intro.html"><a href="intro.html#repro"><i class="fa fa-check"></i><b>1.5</b> Reproducibility</a><ul>
<li class="chapter" data-level="1.5.1" data-path="intro.html"><a href="intro.html#need-for-reproducibility"><i class="fa fa-check"></i><b>1.5.1</b> Need for Reproducibility</a></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html#example-1"><i class="fa fa-check"></i>Example 1</a></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html#example-2"><i class="fa fa-check"></i>Example 2</a></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html#example-3"><i class="fa fa-check"></i>Example 3</a></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html#example-4"><i class="fa fa-check"></i>Example 4</a></li>
<li class="chapter" data-level="1.5.2" data-path="intro.html"><a href="intro.html#the-reproducible-data-analysis-process"><i class="fa fa-check"></i><b>1.5.2</b> The reproducible data analysis process</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="intro.html"><a href="intro.html#data-examples"><i class="fa fa-check"></i><b>1.6</b> Data Examples</a><ul>
<li class="chapter" data-level="1.6.1" data-path="intro.html"><a href="intro.html#college-rankings-systems"><i class="fa fa-check"></i><b>1.6.1</b> College Rankings Systems</a></li>
<li class="chapter" data-level="1.6.2" data-path="intro.html"><a href="intro.html#trump-and-twitter"><i class="fa fa-check"></i><b>1.6.2</b> Trump and Twitter</a></li>
<li class="chapter" data-level="1.6.3" data-path="intro.html"><a href="intro.html#can-twitter-predict-election-results"><i class="fa fa-check"></i><b>1.6.3</b> Can Twitter Predict Election Results?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="visualization.html"><a href="visualization.html"><i class="fa fa-check"></i><b>2</b> Visualization</a><ul>
<li class="chapter" data-level="2.1" data-path="visualization.html"><a href="visualization.html#Sep10"><i class="fa fa-check"></i><b>2.1</b> 9/10/19 Agenda</a></li>
<li class="chapter" data-level="2.2" data-path="visualization.html"><a href="visualization.html#examples"><i class="fa fa-check"></i><b>2.2</b> Examples</a><ul>
<li class="chapter" data-level="2.2.1" data-path="visualization.html"><a href="visualization.html#cholera-via-tufte"><i class="fa fa-check"></i><b>2.2.1</b> Cholera via Tufte</a></li>
<li class="chapter" data-level="2.2.2" data-path="visualization.html"><a href="visualization.html#challenger-via-tufte"><i class="fa fa-check"></i><b>2.2.2</b> Challenger via Tufte</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="visualization.html"><a href="visualization.html#thoughts"><i class="fa fa-check"></i><b>2.3</b> Thoughts on Plotting</a><ul>
<li class="chapter" data-level="2.3.1" data-path="visualization.html"><a href="visualization.html#advice"><i class="fa fa-check"></i><b>2.3.1</b> Advice</a></li>
<li class="chapter" data-level="2.3.2" data-path="visualization.html"><a href="visualization.html#an-example-from-information-is-beautiful"><i class="fa fa-check"></i><b>2.3.2</b> An example from Information is Beautiful</a></li>
<li class="chapter" data-level="2.3.3" data-path="visualization.html"><a href="visualization.html#assessing-graphics-and-other-analyses"><i class="fa fa-check"></i><b>2.3.3</b> Assessing Graphics (and Other Analyses)</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="visualization.html"><a href="visualization.html#Sep12"><i class="fa fa-check"></i><b>2.4</b> 9/12/19 Agenda</a></li>
<li class="chapter" data-level="2.5" data-path="visualization.html"><a href="visualization.html#deconstruct"><i class="fa fa-check"></i><b>2.5</b> Deconstructing a graph</a><ul>
<li class="chapter" data-level="2.5.1" data-path="visualization.html"><a href="visualization.html#gg"><i class="fa fa-check"></i><b>2.5.1</b> The Grammar of Graphics (<code>gg</code>)</a></li>
<li class="chapter" data-level="2.5.2" data-path="visualization.html"><a href="visualization.html#ggplot2"><i class="fa fa-check"></i><b>2.5.2</b> <code>ggplot2</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="wrang.html"><a href="wrang.html"><i class="fa fa-check"></i><b>3</b> Data Wrangling</a><ul>
<li class="chapter" data-level="3.1" data-path="wrang.html"><a href="wrang.html#Sep17"><i class="fa fa-check"></i><b>3.1</b> 9/17/19 Agenda</a></li>
<li class="chapter" data-level="3.2" data-path="wrang.html"><a href="wrang.html#datastruc"><i class="fa fa-check"></i><b>3.2</b> Structure of Data</a><ul>
<li class="chapter" data-level="3.2.1" data-path="wrang.html"><a href="wrang.html#building-tidy-data"><i class="fa fa-check"></i><b>3.2.1</b> Building Tidy Data</a></li>
<li class="chapter" data-level="3.2.2" data-path="wrang.html"><a href="wrang.html#examples-of-chaining"><i class="fa fa-check"></i><b>3.2.2</b> Examples of Chaining</a></li>
<li class="chapter" data-level="3.2.3" data-path="wrang.html"><a href="wrang.html#data-verbs-on-single-data-frames"><i class="fa fa-check"></i><b>3.2.3</b> Data Verbs (on single data frames)</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="wrang.html"><a href="wrang.html#r-examples-basic-verbs"><i class="fa fa-check"></i><b>3.3</b> R examples, basic verbs</a><ul>
<li class="chapter" data-level="3.3.1" data-path="wrang.html"><a href="wrang.html#datasets"><i class="fa fa-check"></i><b>3.3.1</b> Datasets</a></li>
<li class="chapter" data-level="3.3.2" data-path="wrang.html"><a href="wrang.html#examples-of-chaining-1"><i class="fa fa-check"></i><b>3.3.2</b> Examples of Chaining</a></li>
<li class="chapter" data-level="3.3.3" data-path="wrang.html"><a href="wrang.html#data-verbs"><i class="fa fa-check"></i><b>3.3.3</b> Data Verbs</a></li>
<li class="chapter" data-level="3.3.4" data-path="wrang.html"><a href="wrang.html#summarize-and-group_by"><i class="fa fa-check"></i><b>3.3.4</b> <code>summarize</code> and <code>group_by</code></a></li>
<li class="chapter" data-level="3.3.5" data-path="wrang.html"><a href="wrang.html#babynames"><i class="fa fa-check"></i><b>3.3.5</b> babynames</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="wrang.html"><a href="wrang.html#Sep19"><i class="fa fa-check"></i><b>3.4</b> 9/19/19 Agenda</a></li>
<li class="chapter" data-level="3.5" data-path="wrang.html"><a href="wrang.html#highverb"><i class="fa fa-check"></i><b>3.5</b> Higher Level Data Verbs</a></li>
<li class="chapter" data-level="3.6" data-path="wrang.html"><a href="wrang.html#r-examples-higher-level-verbs"><i class="fa fa-check"></i><b>3.6</b> R examples, higher level verbs</a><ul>
<li class="chapter" data-level="3.6.1" data-path="wrang.html"><a href="wrang.html#pivot_longer"><i class="fa fa-check"></i><b>3.6.1</b> <code>pivot_longer</code></a></li>
<li class="chapter" data-level="3.6.2" data-path="wrang.html"><a href="wrang.html#pivot_wider"><i class="fa fa-check"></i><b>3.6.2</b> <code>pivot_wider</code></a></li>
<li class="chapter" data-level="3.6.3" data-path="wrang.html"><a href="wrang.html#join-use-join-to-merge-two-datasets"><i class="fa fa-check"></i><b>3.6.3</b> <code>join</code> (use <code>join</code> to <strong>merge</strong> two datasets)</a></li>
<li class="chapter" data-level="3.6.4" data-path="wrang.html"><a href="wrang.html#lubridate"><i class="fa fa-check"></i><b>3.6.4</b> <code>lubridate</code></a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="wrang.html"><a href="wrang.html#reprex"><i class="fa fa-check"></i><b>3.7</b> <code>reprex</code></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sims.html"><a href="sims.html"><i class="fa fa-check"></i><b>4</b> Simulating</a><ul>
<li class="chapter" data-level="4.1" data-path="sims.html"><a href="sims.html#Sep24"><i class="fa fa-check"></i><b>4.1</b> 9/24/19 Agenda</a></li>
<li class="chapter" data-level="4.2" data-path="sims.html"><a href="sims.html#simmodels"><i class="fa fa-check"></i><b>4.2</b> Simulating Complicated Models</a><ul>
<li class="chapter" data-level="4.2.1" data-path="sims.html"><a href="sims.html#goals-of-simulating-complicated-models"><i class="fa fa-check"></i><b>4.2.1</b> Goals of Simulating Complicated Models</a></li>
<li class="chapter" data-level="4.2.2" data-path="sims.html"><a href="sims.html#examples-of-pigs-and-blackjack"><i class="fa fa-check"></i><b>4.2.2</b> Examples of Pigs and Blackjack</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="sims.html"><a href="sims.html#Sep26"><i class="fa fa-check"></i><b>4.3</b> 9/26/19 Agenda</a></li>
<li class="chapter" data-level="4.4" data-path="sims.html"><a href="sims.html#simsens"><i class="fa fa-check"></i><b>4.4</b> Simulating to Assess Sensitivity</a><ul>
<li class="chapter" data-level="4.4.1" data-path="sims.html"><a href="sims.html#biasmodels"><i class="fa fa-check"></i><b>4.4.1</b> Bias in Models</a></li>
<li class="chapter" data-level="4.4.2" data-path="sims.html"><a href="sims.html#technical-conditions"><i class="fa fa-check"></i><b>4.4.2</b> Technical Conditions</a></li>
<li class="chapter" data-level="4.4.3" data-path="sims.html"><a href="sims.html#generating-random-numbers"><i class="fa fa-check"></i><b>4.4.3</b> Generating random numbers</a></li>
<li class="chapter" data-level="4.4.4" data-path="sims.html"><a href="sims.html#generating-other-rvs-the-inverse-transform-method"><i class="fa fa-check"></i><b>4.4.4</b> Generating other RVs: <strong>The Inverse Transform Method</strong></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="permschp.html"><a href="permschp.html"><i class="fa fa-check"></i><b>5</b> Permutation Tests</a><ul>
<li class="chapter" data-level="5.1" data-path="permschp.html"><a href="permschp.html#Oct1"><i class="fa fa-check"></i><b>5.1</b> 10/1/19 Agenda</a></li>
<li class="chapter" data-level="5.2" data-path="permschp.html"><a href="permschp.html#algs"><i class="fa fa-check"></i><b>5.2</b> Inference Algorithms</a><ul>
<li class="chapter" data-level="5.2.1" data-path="permschp.html"><a href="permschp.html#hypothesis-test-algorithm"><i class="fa fa-check"></i><b>5.2.1</b> Hypothesis Test Algorithm</a></li>
<li class="chapter" data-level="5.2.2" data-path="permschp.html"><a href="permschp.html#permutation-tests-algorithm"><i class="fa fa-check"></i><b>5.2.2</b> Permutation Tests Algorithm</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="permschp.html"><a href="permschp.html#Oct3"><i class="fa fa-check"></i><b>5.3</b> 10/3/19 Agenda</a></li>
<li class="chapter" data-level="5.4" data-path="permschp.html"><a href="permschp.html#perms"><i class="fa fa-check"></i><b>5.4</b> Permutation tests in practice</a><ul>
<li class="chapter" data-level="5.4.1" data-path="permschp.html"><a href="permschp.html#permutation-vs.randomization-tests"><i class="fa fa-check"></i><b>5.4.1</b> Permutation vs. Randomization Tests</a></li>
<li class="chapter" data-level="5.4.2" data-path="permschp.html"><a href="permschp.html#ci-from-permutation-tests"><i class="fa fa-check"></i><b>5.4.2</b> CI from Permutation Tests</a></li>
<li class="chapter" data-level="5.4.3" data-path="permschp.html"><a href="permschp.html#randomization-example"><i class="fa fa-check"></i><b>5.4.3</b> Randomization Example</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="permschp.html"><a href="permschp.html#r-examples"><i class="fa fa-check"></i><b>5.5</b> R examples</a><ul>
<li class="chapter" data-level="5.5.1" data-path="permschp.html"><a href="permschp.html#cloud-seeding-two-sample-test-computationally-very-difficult-to-do-a-randomization-test"><i class="fa fa-check"></i><b>5.5.1</b> Cloud Seeding (Two sample test – computationally very difficult to do a randomization test)</a></li>
<li class="chapter" data-level="5.5.2" data-path="permschp.html"><a href="permschp.html#macnell-teaching-evaluations-stratified-two-sample-t-test"><i class="fa fa-check"></i><b>5.5.2</b> MacNell Teaching Evaluations (Stratified two-sample t-test)</a></li>
<li class="chapter" data-level="5.5.3" data-path="permschp.html"><a href="permschp.html#income-and-health-f-like-test"><i class="fa fa-check"></i><b>5.5.3</b> Income and Health (F-like test)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="boot.html"><a href="boot.html"><i class="fa fa-check"></i><b>6</b> Bootstrapping</a><ul>
<li class="chapter" data-level="6.1" data-path="boot.html"><a href="boot.html#Oct8"><i class="fa fa-check"></i><b>6.1</b> 10/8/19 Agenda</a></li>
<li class="chapter" data-level="6.2" data-path="boot.html"><a href="boot.html#introduction"><i class="fa fa-check"></i><b>6.2</b> Introduction</a></li>
<li class="chapter" data-level="6.3" data-path="boot.html"><a href="boot.html#BSnotation"><i class="fa fa-check"></i><b>6.3</b> Basics &amp; Notation</a><ul>
<li class="chapter" data-level="6.3.1" data-path="boot.html"><a href="boot.html#the-plug-in-principle"><i class="fa fa-check"></i><b>6.3.1</b> The Plug-in Principle</a></li>
<li class="chapter" data-level="6.3.2" data-path="boot.html"><a href="boot.html#the-bootstrap-idea"><i class="fa fa-check"></i><b>6.3.2</b> The Bootstrap Idea</a></li>
<li class="chapter" data-level="6.3.3" data-path="boot.html"><a href="boot.html#bootstrap-procedure"><i class="fa fa-check"></i><b>6.3.3</b> Bootstrap Procedure</a></li>
<li class="chapter" data-level="6.3.4" data-path="boot.html"><a href="boot.html#bootstrap-notation"><i class="fa fa-check"></i><b>6.3.4</b> Bootstrap Notation</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="boot.html"><a href="boot.html#Oct10"><i class="fa fa-check"></i><b>6.4</b> 10/10/19 Agenda</a></li>
<li class="chapter" data-level="6.5" data-path="boot.html"><a href="boot.html#BSCI"><i class="fa fa-check"></i><b>6.5</b> Bootstrap Confidence Intervals</a><ul>
<li class="chapter" data-level="6.5.1" data-path="boot.html"><a href="boot.html#normal-standard-ci-with-bs-se-typenorm"><i class="fa fa-check"></i><b>6.5.1</b> Normal (standard) CI with BS SE: <code>type=&quot;norm&quot;</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="boot.html"><a href="boot.html#bootstrap-t-confidence-intervals-typestud"><i class="fa fa-check"></i><b>6.5.2</b> Bootstrap-t Confidence Intervals: <code>type=&quot;stud&quot;</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="boot.html"><a href="boot.html#percentile-confidence-intervals-typeperc"><i class="fa fa-check"></i><b>6.5.3</b> Percentile Confidence Intervals: <code>type=&quot;perc&quot;</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="boot.html"><a href="boot.html#what-makes-a-ci-good"><i class="fa fa-check"></i><b>6.5.4</b> What makes a CI good?</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="boot.html"><a href="boot.html#r-example-heroin"><i class="fa fa-check"></i><b>6.6</b> R example: Heroin</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ethics.html"><a href="ethics.html"><i class="fa fa-check"></i><b>7</b> Ethics</a><ul>
<li class="chapter" data-level="7.1" data-path="ethics.html"><a href="ethics.html#Oct24"><i class="fa fa-check"></i><b>7.1</b> 10/24/19 Agenda</a></li>
<li class="chapter" data-level="7.2" data-path="ethics.html"><a href="ethics.html#doing-data-science"><i class="fa fa-check"></i><b>7.2</b> Doing Data Science</a></li>
<li class="chapter" data-level="7.3" data-path="ethics.html"><a href="ethics.html#graphics"><i class="fa fa-check"></i><b>7.3</b> Graphics</a></li>
<li class="chapter" data-level="7.4" data-path="ethics.html"><a href="ethics.html#p-hacking"><i class="fa fa-check"></i><b>7.4</b> p-hacking</a><ul>
<li class="chapter" data-level="7.4.1" data-path="ethics.html"><a href="ethics.html#multiple-studies"><i class="fa fa-check"></i><b>7.4.1</b> Multiple Studies</a></li>
<li class="chapter" data-level="7.4.2" data-path="ethics.html"><a href="ethics.html#p-values"><i class="fa fa-check"></i><b>7.4.2</b> p-values</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="ethics.html"><a href="ethics.html#human-subjects-research"><i class="fa fa-check"></i><b>7.5</b> Human Subjects Research</a></li>
<li class="chapter" data-level="7.6" data-path="ethics.html"><a href="ethics.html#authorship"><i class="fa fa-check"></i><b>7.6</b> Authorship</a></li>
<li class="chapter" data-level="7.7" data-path="ethics.html"><a href="ethics.html#algorithms"><i class="fa fa-check"></i><b>7.7</b> Algorithms</a><ul>
<li class="chapter" data-level="" data-path="ethics.html"><a href="ethics.html#sentencing"><i class="fa fa-check"></i>Sentencing</a></li>
<li class="chapter" data-level="" data-path="ethics.html"><a href="ethics.html#algorithmic-justice-league"><i class="fa fa-check"></i>Algorithmic Justice League</a></li>
<li class="chapter" data-level="" data-path="ethics.html"><a href="ethics.html#sentiment-analysis"><i class="fa fa-check"></i>Sentiment Analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="ethics.html"><a href="ethics.html#guiding-ethical-principles"><i class="fa fa-check"></i><b>7.8</b> Guiding Ethical Principles</a><ul>
<li class="chapter" data-level="" data-path="ethics.html"><a href="ethics.html#asa-ethical-guidelines-for-statistical-practice"><i class="fa fa-check"></i>ASA Ethical Guidelines for Statistical Practice</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="class.html"><a href="class.html"><i class="fa fa-check"></i><b>8</b> Classification</a><ul>
<li class="chapter" data-level="8.1" data-path="class.html"><a href="class.html#Oct29"><i class="fa fa-check"></i><b>8.1</b> 10/29/19 Agenda</a></li>
<li class="chapter" data-level="8.2" data-path="class.html"><a href="class.html#cv"><i class="fa fa-check"></i><b>8.2</b> Cross Validation</a><ul>
<li class="chapter" data-level="8.2.1" data-path="class.html"><a href="class.html#bias-variance-trade-off"><i class="fa fa-check"></i><b>8.2.1</b> Bias-variance trade-off</a></li>
<li class="chapter" data-level="8.2.2" data-path="class.html"><a href="class.html#implementing-cross-validation"><i class="fa fa-check"></i><b>8.2.2</b> Implementing Cross Validation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="class.html"><a href="class.html#knn"><i class="fa fa-check"></i><b>8.3</b> <span class="math inline">\(k\)</span>-Nearest Neighbors</a><ul>
<li class="chapter" data-level="8.3.1" data-path="class.html"><a href="class.html#k-nn-algorithm"><i class="fa fa-check"></i><b>8.3.1</b> <span class="math inline">\(k\)</span>-NN algorithm</a></li>
<li class="chapter" data-level="8.3.2" data-path="class.html"><a href="class.html#r-knn-example"><i class="fa fa-check"></i><b>8.3.2</b> R knn Example</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="class.html"><a href="class.html#Oct31"><i class="fa fa-check"></i><b>8.4</b> 10/31/19 Agenda</a></li>
<li class="chapter" data-level="8.5" data-path="class.html"><a href="class.html#cart"><i class="fa fa-check"></i><b>8.5</b> CART</a><ul>
<li class="chapter" data-level="8.5.1" data-path="class.html"><a href="class.html#cart-algorithm"><i class="fa fa-check"></i><b>8.5.1</b> CART algorithm</a></li>
<li class="chapter" data-level="8.5.2" data-path="class.html"><a href="class.html#r-cart-example"><i class="fa fa-check"></i><b>8.5.2</b> R CART Example</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="class.html"><a href="class.html#Nov5"><i class="fa fa-check"></i><b>8.6</b> 11/5/19 Agenda</a></li>
<li class="chapter" data-level="8.7" data-path="class.html"><a href="class.html#bagging"><i class="fa fa-check"></i><b>8.7</b> Bagging</a><ul>
<li class="chapter" data-level="8.7.1" data-path="class.html"><a href="class.html#bagging-algorithm"><i class="fa fa-check"></i><b>8.7.1</b> Bagging algorithm</a></li>
<li class="chapter" data-level="8.7.2" data-path="class.html"><a href="class.html#out-of-bag-oob-error-rate"><i class="fa fa-check"></i><b>8.7.2</b> Out Of Bag (OOB) error rate</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="class.html"><a href="class.html#Nov7"><i class="fa fa-check"></i><b>8.8</b> 11/7/19 Agenda</a></li>
<li class="chapter" data-level="8.9" data-path="class.html"><a href="class.html#rf"><i class="fa fa-check"></i><b>8.9</b> Random Forests</a><ul>
<li class="chapter" data-level="8.9.1" data-path="class.html"><a href="class.html#random-forest-algorithm"><i class="fa fa-check"></i><b>8.9.1</b> Random Forest algorithm</a></li>
<li class="chapter" data-level="8.9.2" data-path="class.html"><a href="class.html#r-rf-example"><i class="fa fa-check"></i><b>8.9.2</b> R RF Example</a></li>
</ul></li>
<li class="chapter" data-level="8.10" data-path="class.html"><a href="class.html#model-choices"><i class="fa fa-check"></i><b>8.10</b> Model Choices</a></li>
<li class="chapter" data-level="8.11" data-path="class.html"><a href="class.html#Nov12"><i class="fa fa-check"></i><b>8.11</b> 11/12/19 Agenda</a></li>
<li class="chapter" data-level="8.12" data-path="class.html"><a href="class.html#support-vector-machines"><i class="fa fa-check"></i><b>8.12</b> Support Vector Machines</a><ul>
<li class="chapter" data-level="8.12.1" data-path="class.html"><a href="class.html#Nov14"><i class="fa fa-check"></i><b>8.12.1</b> 11/14/19 Agenda</a></li>
<li class="chapter" data-level="8.12.2" data-path="class.html"><a href="class.html#notlinsvm"><i class="fa fa-check"></i><b>8.12.2</b> Not Linearly Separable</a></li>
<li class="chapter" data-level="8.12.3" data-path="class.html"><a href="class.html#Nov19"><i class="fa fa-check"></i><b>8.12.3</b> 11/19/19 Agenda</a></li>
<li class="chapter" data-level="8.12.4" data-path="class.html"><a href="class.html#support-vector-machine-algorithm"><i class="fa fa-check"></i><b>8.12.4</b> Support Vector Machine algorithm</a></li>
<li class="chapter" data-level="8.12.5" data-path="class.html"><a href="class.html#classifying-more-than-one-group"><i class="fa fa-check"></i><b>8.12.5</b> Classifying more than one group</a></li>
<li class="chapter" data-level="8.12.6" data-path="class.html"><a href="class.html#r-svm-example"><i class="fa fa-check"></i><b>8.12.6</b> R SVM Example</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="unsup.html"><a href="unsup.html"><i class="fa fa-check"></i><b>9</b> Unsupervised Methods</a><ul>
<li class="chapter" data-level="9.1" data-path="unsup.html"><a href="unsup.html#Nov21"><i class="fa fa-check"></i><b>9.1</b> 11/21/19 Agenda</a></li>
<li class="chapter" data-level="9.2" data-path="unsup.html"><a href="unsup.html#latent-dirichlet-allocation"><i class="fa fa-check"></i><b>9.2</b> Latent Dirichlet Allocation</a></li>
<li class="chapter" data-level="9.3" data-path="unsup.html"><a href="unsup.html#dissimilarities"><i class="fa fa-check"></i><b>9.3</b> Dissimilarities</a></li>
<li class="chapter" data-level="9.4" data-path="unsup.html"><a href="unsup.html#hier"><i class="fa fa-check"></i><b>9.4</b> Hierarchical Clustering</a></li>
<li class="chapter" data-level="9.5" data-path="unsup.html"><a href="unsup.html#part"><i class="fa fa-check"></i><b>9.5</b> Partitioning Clustering</a><ul>
<li class="chapter" data-level="9.5.1" data-path="unsup.html"><a href="unsup.html#k-means-clustering"><i class="fa fa-check"></i><b>9.5.1</b> <span class="math inline">\(k\)</span>-means Clustering</a></li>
<li class="chapter" data-level="9.5.2" data-path="unsup.html"><a href="unsup.html#partitioning-around-medoids"><i class="fa fa-check"></i><b>9.5.2</b> Partitioning Around Medoids</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="unsup.html"><a href="unsup.html#evaluation-metrics"><i class="fa fa-check"></i><b>9.6</b> Evaluation Metrics</a></li>
<li class="chapter" data-level="9.7" data-path="unsup.html"><a href="unsup.html#em-algorithm"><i class="fa fa-check"></i><b>9.7</b> EM algorithm</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="misc.html"><a href="misc.html"><i class="fa fa-check"></i><b>10</b> Misc</a><ul>
<li class="chapter" data-level="10.1" data-path="misc.html"><a href="misc.html#Nov26"><i class="fa fa-check"></i><b>10.1</b> 11/26/19 Agenda</a></li>
<li class="chapter" data-level="10.2" data-path="misc.html"><a href="misc.html#api"><i class="fa fa-check"></i><b>10.2</b> API</a><ul>
<li class="chapter" data-level="10.2.1" data-path="misc.html"><a href="misc.html#authenticating"><i class="fa fa-check"></i><b>10.2.1</b> Authenticating</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="misc.html"><a href="misc.html#parallel-computing"><i class="fa fa-check"></i><b>10.3</b> Parallel Computing</a><ul>
<li class="chapter" data-level="10.3.1" data-path="misc.html"><a href="misc.html#spark-and-sparklyr"><i class="fa fa-check"></i><b>10.3.1</b> Spark and <code>sparklyr</code></a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="misc.html"><a href="misc.html#cloud-computing"><i class="fa fa-check"></i><b>10.4</b> Cloud Computing</a></li>
<li class="chapter" data-level="10.5" data-path="misc.html"><a href="misc.html#reticulate"><i class="fa fa-check"></i><b>10.5</b> <code>reticulate</code></a><ul>
<li class="chapter" data-level="" data-path="misc.html"><a href="misc.html#connect-to-python-within-rstudio"><i class="fa fa-check"></i>Connect to Python within RStudio</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="misc.html"><a href="misc.html#sql-in-r"><i class="fa fa-check"></i><b>10.6</b> SQL (in R)</a></li>
<li class="chapter" data-level="10.7" data-path="misc.html"><a href="misc.html#Dec10"><i class="fa fa-check"></i><b>10.7</b> 12/10/19 Agenda</a></li>
<li class="chapter" data-level="10.8" data-path="misc.html"><a href="misc.html#regexpr"><i class="fa fa-check"></i><b>10.8</b> Regular Expressions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="compstat.html"><a href="compstat.html"><i class="fa fa-check"></i><b>11</b> Computational Statistics</a><ul>
<li class="chapter" data-level="11.1" data-path="compstat.html"><a href="compstat.html#wearing-a-statistics-hat"><i class="fa fa-check"></i><b>11.1</b> Wearing a Statistics Hat</a></li>
<li class="chapter" data-level="11.2" data-path="compstat.html"><a href="compstat.html#computational-statistics"><i class="fa fa-check"></i><b>11.2</b> Computational Statistics</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://st47s.com/Math154" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Statistics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="misc" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Misc</h1>
<div id="Nov26" class="section level2">
<h2><span class="header-section-number">10.1</span> 11/26/19 Agenda</h2>
<ol style="list-style-type: decimal">
<li>API / authenticating</li>
<li>parallel computing</li>
<li>cloud computing</li>
<li><code>reticulate</code> (Python in R!)</li>
<li>SQL</li>
</ol>
</div>
<div id="api" class="section level2">
<h2><span class="header-section-number">10.2</span> API</h2>
<div style="float:right;position: relative; top: -30px;">
<div class="figure" style="text-align: right"><span id="fig:unnamed-chunk-2"></span>
<img src="figs/api_xkcd.png" alt="xkcd, https://xkcd.com/1481/" width="60%" />
<p class="caption">
Figure 1.2: xkcd, <a href="https://xkcd.com/1481/" class="uri">https://xkcd.com/1481/</a>
</p>
</div>
</div>
<p>What is an API? (Application Programming Interface)</p>
<p>Think of an API as a restaurant menu. The menu provides a list of what the restaurant has to offer, and you order off the menu by choosing the dish that you want. After you order, the restaurant figures out how to bring the food from the kitchen to your table in the way that you’ve specified.</p>
<p>An API is an intermediary that allows two applications to talk to one another. It is not the database or the server, instead it is the <em>code</em> that allows communication.</p>
<div id="examples-of-apis" class="section level4 unnumbered">
<h4>Examples of APIs</h4>
<ul>
<li><p>When you use an app on your phone, the app connects to the internet and sends information to a server somewhere. The server retrieves the data, interprets is, does what it does, and sends it back to you. The application which takes the data from the server and presents it to you in a readable way is an API.</p></li>
<li><p>Let’s say you are booking a flight on United. You choose all the details, you interact with the airline’s website. BUT INSTEAD, what if you are interacting with a software like Expedia? Then Expedia has to talk to United’s API to get all the information about available flights, costs, seats, etc.</p></li>
<li><p>If you’ve ever been to a third party site and clicked on “Share on Facebook” or “Share on Twitter” your third party site is communicating with the Facebook API or the Twitter API.</p></li>
<li><p>You sign up to go to a concert, and StubHub asks whether you want to add the concert to your Google calendar. StubHub needs to talk to Google via Google’s API.</p></li>
<li><p>What if you want some Twitter data? How might you get it? Well, you could email Twitter and ask someone for it. <strong>Instead</strong> Twitter provides information about how their data is stored, and allows you to query their data in an automated way.</p></li>
</ul>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-3"></span>
<img src="figs/API_weather.png" alt="Image taken from https://rigor.com/blog/what-is-an-api-a-brief-intro" width="60%" />
<p class="caption">
Figure 1.3: Image taken from <a href="https://rigor.com/blog/what-is-an-api-a-brief-intro" class="uri">https://rigor.com/blog/what-is-an-api-a-brief-intro</a>
</p>
</div>
</div>
<div id="authenticating" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Authenticating</h3>
<ul>
<li><strong>Authenticating</strong> is stating who you are</li>
<li><strong>Authorization</strong> is asking for access to a resources (and happens after authentication)</li>
</ul>
<p>DO NOT post your credentials and keys to a public GitHub repo!!</p>
<p>In almost all cases, in order to communicate with an API, you must tell the API who you are and that you should have access to the information the API is providing.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-4"></span>
<img src="figs/auth.png" alt="Image taken from https://blog.restcase.com/restful-api-authentication-basics/" width="60%" />
<p class="caption">
Figure 1.4: Image taken from <a href="https://blog.restcase.com/restful-api-authentication-basics/" class="uri">https://blog.restcase.com/restful-api-authentication-basics/</a>
</p>
</div>
</div>
</div>
<div id="parallel-computing" class="section level2">
<h2><span class="header-section-number">10.3</span> Parallel Computing</h2>
<p>(Taken from the Teach Data Science blog: <a href="https://teachdatascience.com/parallel/" class="uri">https://teachdatascience.com/parallel/</a>)</p>
<p>To demonstrate what parallel computing is, we’ll perform tasks that are <strong>embarrassingly parallel</strong> which means there is no dependency or communication between the parallel tasks. Again, parallel computing can be powerful in ways that link computational tasks in complicated ways. But we believe that as a <em>first</em> pass at teaching parallel computing, we should teach the parallel structure before bringing in dependence across the parallel tasks. Examples of <a href="https://en.wikipedia.org/wiki/Embarrassingly_parallel">embarrassingly parallel</a> algorithms include: Monte Carlo analysis, bootstrapping, growing trees for Random Forests, <code>group_by</code> analyses, and cross-validation. Additionally, data science methods increasingly use <a href="https://en.wikipedia.org/wiki/Randomized_algorithm">randomized algorithms</a> which can often be written in parallel.</p>
<p>Indeed, it isn’t always easy to know when to use a parallel construction. Because of existing overhead processes (e.g., copying data across many threads, bring results together, etc.) an algorithm run on 10 parallel strands will not reduce an original (non-parallel) run time by 10-fold. Figuring out when a parallel implementation is appropriate is beyond the scope of this blog but should be carefully considered before embarking on large projects.</p>
<div id="some-parallel-examples" class="section level4 unnumbered">
<h4>Some parallel examples</h4>
<p>Before running code in parallel, it is valuable to know how many cores your computer has to work with. Note that the <code>detectCores</code> function will provide information about the specific device you are using (<code>logical = FALSE</code> tells you only the physical cores which is likely what you want). Note that after <code>makeCluster</code> the separate threads have information. [The argument <code>setup_strategy = &quot;sequential&quot;</code> is a fix for working with R 4.0+.] After <code>stopCluster</code>, the code is no longer connecting to the cluster structure.</p>
<div class="sourceCode" id="cb509"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb509-1" data-line-number="1"><span class="kw">library</span>(parallel)</a>
<a class="sourceLine" id="cb509-2" data-line-number="2">P &lt;-<span class="st"> </span><span class="kw">detectCores</span>(<span class="dt">logical=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb509-3" data-line-number="3">P</a></code></pre></div>
<pre><code>## [1] 8</code></pre>
<div class="sourceCode" id="cb511"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb511-1" data-line-number="1">cl &lt;-<span class="st"> </span><span class="kw">makeCluster</span>(P, <span class="dt">setup_strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb511-2" data-line-number="2">cl[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## node of a socket cluster on host &#39;localhost&#39; with pid 75490</code></pre>
<div class="sourceCode" id="cb513"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb513-1" data-line-number="1"><span class="kw">stopCluster</span>(cl)</a>
<a class="sourceLine" id="cb513-2" data-line-number="2">cl[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## Error in summary.connection(connection): invalid connection</code></pre>
</div>
<div id="embarrassingly-embarrassing-example" class="section level4 unnumbered">
<h4>Embarrassingly embarrassing example</h4>
<p>In the example below, we generate some Cauchy data and find the max of each sample. Note that for the current device there are 8 cores, so the process will happen 100/P = 12.5 times on each core. The second argument of <code>clusterApply</code> is a sequence of numbers that gets passed to each worker as the (first) argument of <code>func1</code>. Below, I’ve specified that the value 50 (number of reps) should be passed separately to 100 different workers.</p>
<div class="sourceCode" id="cb515"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb515-1" data-line-number="1">W &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb515-2" data-line-number="2">P &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">detectCores</span>(<span class="dt">logical=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb515-3" data-line-number="3">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(P, <span class="dt">setup_strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb515-4" data-line-number="4"></a>
<a class="sourceLine" id="cb515-5" data-line-number="5">func1 &lt;-<span class="st"> </span><span class="cf">function</span>(reps){</a>
<a class="sourceLine" id="cb515-6" data-line-number="6">  <span class="kw">max</span>(<span class="kw">rcauchy</span>(reps))</a>
<a class="sourceLine" id="cb515-7" data-line-number="7">}</a>
<a class="sourceLine" id="cb515-8" data-line-number="8"></a>
<a class="sourceLine" id="cb515-9" data-line-number="9"></a>
<a class="sourceLine" id="cb515-10" data-line-number="10"><span class="kw">clusterApply</span>(cl, <span class="kw">rep</span>(<span class="dv">50</span>,W), <span class="dt">fun =</span> func1) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">3</span>)</a></code></pre></div>
<pre><code>## [[1]]
## [1] 36.44004
## 
## [[2]]
## [1] 41.72103
## 
## [[3]]
## [1] 74.68404</code></pre>
<div class="sourceCode" id="cb517"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb517-1" data-line-number="1"><span class="kw">stopCluster</span>(cl)</a></code></pre></div>
<p>There are many R functions which implement parallel processing. For example, the same code from above can be processed using <code>foreach</code>.</p>
<div class="sourceCode" id="cb518"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb518-1" data-line-number="1"><span class="kw">library</span>(doParallel)</a>
<a class="sourceLine" id="cb518-2" data-line-number="2">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(P, <span class="dt">setup_strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb518-3" data-line-number="3"></a>
<a class="sourceLine" id="cb518-4" data-line-number="4">doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(cl)</a>
<a class="sourceLine" id="cb518-5" data-line-number="5"><span class="kw">foreach</span>(<span class="dt">reps =</span> <span class="kw">rep</span>(<span class="dv">50</span>, <span class="dv">100</span>), <span class="dt">.combine =</span> <span class="st">&#39;c&#39;</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb518-6" data-line-number="6">  <span class="kw">max</span>(<span class="kw">rcauchy</span>(reps))</a>
<a class="sourceLine" id="cb518-7" data-line-number="7">       } <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">3</span>)</a></code></pre></div>
<pre><code>## [1]  43.14389 274.16503  19.45930</code></pre>
<div class="sourceCode" id="cb520"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb520-1" data-line-number="1"><span class="kw">stopCluster</span>(cl)</a></code></pre></div>
</div>
<div id="example-bootstrapping" class="section level4 unnumbered">
<h4>Example bootstrapping</h4>
<p>A slightly less embarrassingly parallel example comes with bootstrapping. Below we have used parallel implementation to bootstrap the mean of the iris data petal length (Virginica only).</p>
<div class="sourceCode" id="cb521"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb521-1" data-line-number="1"><span class="kw">data</span>(iris)</a>
<a class="sourceLine" id="cb521-2" data-line-number="2"></a>
<a class="sourceLine" id="cb521-3" data-line-number="3">iris_bs &lt;-<span class="st"> </span>iris <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb521-4" data-line-number="4"><span class="st">  </span><span class="kw">filter</span>(Species <span class="op">==</span><span class="st"> &quot;virginica&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb521-5" data-line-number="5"><span class="st">  </span><span class="kw">select</span>(Petal.Length)</a></code></pre></div>
<div class="sourceCode" id="cb522"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb522-1" data-line-number="1">cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">makeCluster</span>(P, <span class="dt">setup_strategy =</span> <span class="st">&quot;sequential&quot;</span>)</a>
<a class="sourceLine" id="cb522-2" data-line-number="2"></a>
<a class="sourceLine" id="cb522-3" data-line-number="3">doParallel<span class="op">::</span><span class="kw">registerDoParallel</span>(cl)</a>
<a class="sourceLine" id="cb522-4" data-line-number="4">bsmean_PL &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>, <span class="dt">.combine =</span> <span class="st">&#39;c&#39;</span>) <span class="op">%dopar%</span><span class="st"> </span>{</a>
<a class="sourceLine" id="cb522-5" data-line-number="5">  <span class="kw">mean</span>(<span class="kw">sample</span>(iris_bs<span class="op">$</span>Petal.Length, <span class="dt">replace =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb522-6" data-line-number="6">}</a>
<a class="sourceLine" id="cb522-7" data-line-number="7">bootstrap &lt;-<span class="st"> </span><span class="kw">tibble</span>(bsmean_PL)</a>
<a class="sourceLine" id="cb522-8" data-line-number="8"><span class="kw">stopCluster</span>(cl)</a>
<a class="sourceLine" id="cb522-9" data-line-number="9"></a>
<a class="sourceLine" id="cb522-10" data-line-number="10"><span class="kw">ggplot</span>(bootstrap, <span class="kw">aes</span>(<span class="dt">x =</span> bsmean_PL)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">25</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Histogram of 100 Bootstrapped Means using foreach&quot;</span>)</a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-9-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="spark-and-sparklyr" class="section level3">
<h3><span class="header-section-number">10.3.1</span> Spark and <code>sparklyr</code></h3>
<p>Some of you may be familiar with <a href="http://spark.apache.org/">Apache Spark</a> which is an open-source product for distributed cluster-computing. You may want to learn more about its capabilities, including scheduling workflow, dispatching tasks, and consolidating end results. While incredibly powerful, there has historically been a steep learning curve to getting R to work smoothly with a Spark connection. Recently, RStudio has come out with a new package <a href="https://spark.rstudio.com/"><code>sparklyr</code></a> which integrates R and Spark seamlessly. Note that in the example below, we’ve set up a local connection just for the purposes of the example. For your work, you may want to connect to a cluster or cloud space with many cores.</p>
<p>The <a href="https://spark.rstudio.com/">RStudio <code>sparklyr</code> webpage</a> provides a plethora of good examples demonstrating the sophistication and power of the technology. <code>sparklyr</code> has particularly strong connections to the suite of <code>tidyverse</code> functions. Indeed, the power of <code>sparklyr</code> is more about distributing the computing than about parallelizing it. For example, with <code>sparklyr</code> the computations are delayed until you need the results. Additionally, Spark is doing the heavy lifting and only at the very end (when your results are called) do you need to worry about the size of the table, results, or computational space. The example below repeats the bootstrapping work that was done previously.</p>
<p>Note, it is important to look at your data structures and variables names. For example, when copying the local dataframe <code>iris_samps</code> to the remote data source called <code>iris_samps_tbl</code>, the variable <code>Petal.Length</code> was changed to <code>Petal_Length</code>.</p>
<div class="sourceCode" id="cb523"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb523-1" data-line-number="1"><span class="kw">library</span>(sparklyr)</a>
<a class="sourceLine" id="cb523-2" data-line-number="2"><span class="kw">spark_install</span>()</a>
<a class="sourceLine" id="cb523-3" data-line-number="3"></a>
<a class="sourceLine" id="cb523-4" data-line-number="4">sc &lt;-<span class="st"> </span><span class="kw">spark_connect</span>(<span class="dt">master =</span> <span class="st">&quot;local&quot;</span>)</a>
<a class="sourceLine" id="cb523-5" data-line-number="5"></a>
<a class="sourceLine" id="cb523-6" data-line-number="6">n_sim =<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb523-7" data-line-number="7">iris_samps &lt;-<span class="st"> </span>iris <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(Species <span class="op">==</span><span class="st"> &quot;virginica&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb523-8" data-line-number="8"><span class="st">  </span><span class="kw">sapply</span>(rep.int, <span class="dt">times=</span>n_sim) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">cbind</span>(<span class="dt">replicate =</span> <span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span>n_sim, <span class="dt">each =</span> <span class="dv">50</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb523-9" data-line-number="9"><span class="st">  </span><span class="kw">data.frame</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb523-10" data-line-number="10"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">group_by</span>(replicate) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb523-11" data-line-number="11"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">sample_n</span>(<span class="dv">50</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb523-12" data-line-number="12"></a>
<a class="sourceLine" id="cb523-13" data-line-number="13">iris_samps_tbl &lt;-<span class="st"> </span><span class="kw">copy_to</span>(sc, iris_samps)</a>
<a class="sourceLine" id="cb523-14" data-line-number="14"></a>
<a class="sourceLine" id="cb523-15" data-line-number="15">iris_samps_tbl <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb523-16" data-line-number="16"><span class="st">  </span><span class="kw">spark_apply</span>(<span class="cf">function</span>(x) {<span class="kw">mean</span>(x<span class="op">$</span>Petal_Length)}, </a>
<a class="sourceLine" id="cb523-17" data-line-number="17">    <span class="dt">group_by =</span> <span class="st">&quot;replicate&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb523-18" data-line-number="18"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> result)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">20</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Histogram of 100 Bootstrapped Means using sparklyr&quot;</span>)</a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-10-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb524"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb524-1" data-line-number="1"><span class="kw">spark_disconnect</span>(sc)</a></code></pre></div>
<p>For our particular application, the adept reader has probably noticed that the average of a variable using <code>group_by</code> is a very quick and easy task for <code>dplyr</code>. Indeed, the use of <code>sparklyr</code> above is overkill and is presented only as a way to demonstrate using <code>sparklyr</code>. If you are working with big datasets that require large computing infrastructure, the RStudio help pages on <code>sparklyr</code> are fantastic. Additionally, there are many instances of working with <code>Spark</code> in the wild, and you might consider working through someone else’s <code>Spark</code> analysis like this <a href="https://livefreeordichotomize.com/2019/06/04/using_awk_and_r_to_parse_25tb/">fantastic example</a> on splitting up large amounts of raw DNA sequencing to get data for a given genetic location.</p>
<div class="sourceCode" id="cb525"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb525-1" data-line-number="1">iris_samps <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">group_by</span>(replicate) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb525-2" data-line-number="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">summarize</span>(<span class="dt">result =</span> <span class="kw">mean</span>(Petal.Length)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb525-3" data-line-number="3"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> result)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">25</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Histogram of 100 Bootstrapped Means using dplyr&quot;</span>)</a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-11-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>While an introduction to parallel and cloud computing will help you become more adept and less apprehensive about using the tools, there is also a recognition that sufficient background in computer science is needed to be able to fully engage with principles of high performance computing.</p>
<div id="learn-more" class="section level4 unnumbered">
<h4>Learn more</h4>
<ul>
<li>Hana Sevcikova <a href="https://channel9.msdn.com/Events/useR-international-R-User-conferences/useR-International-R-User-2017-Conference/Introduction-to-parallel-computing-with-R">Introduction to parallel computing with R</a> useR 2017 in Brussels, <a href="https://rawgit.com/PPgp/useR2017public/master/tutorial.html">tutorial here</a></li>
<li><code>sparklyr</code> to do <a href="https://blog.rstudio.com/2018/05/14/sparklyr-0-8/">parallel cross-validation</a></li>
<li><a href="https://www.rstudio.com/resources/cheatsheets/" class="uri">https://www.rstudio.com/resources/cheatsheets/</a>
<ul>
<li><a href="https://www.rstudio.com/resources/cheatsheets/#sparklyr" class="uri">https://www.rstudio.com/resources/cheatsheets/#sparklyr</a><br />
</li>
<li><a href="https://github.com/rstudio/cheatsheets/raw/master/parallel_computation.pdf" class="uri">https://github.com/rstudio/cheatsheets/raw/master/parallel_computation.pdf</a></li>
</ul></li>
<li>Great blog <a href="https://thecodeforest.github.io/post/two_flavors_of_parallel_simulation.html">Two Flavors of Parallel Simulation</a> by Mark LeBoeuf comparing different ways to process code in parallel.</li>
</ul>
</div>
</div>
</div>
<div id="cloud-computing" class="section level2">
<h2><span class="header-section-number">10.4</span> Cloud Computing</h2>
<p>A great overview on high performance computing (HPC) what it is and what it isn’t is given here: <a href="https://www.slideshare.net/raamana/high-performance-computing-with-checklist-and-tips-optimal-cluster-usage" class="uri">https://www.slideshare.net/raamana/high-performance-computing-with-checklist-and-tips-optimal-cluster-usage</a></p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-12"></span>
<img src="figs/notHPC.png" alt="Image from Pradeep Redddy Raamana [High performance computing tutorial, with checklist and tips to optimize cluster usage](https://www.slideshare.net/raamana/high-performance-computing-with-checklist-and-tips-optimal-cluster-usage)" width="45%" /><img src="figs/whenHPC.png" alt="Image from Pradeep Redddy Raamana [High performance computing tutorial, with checklist and tips to optimize cluster usage](https://www.slideshare.net/raamana/high-performance-computing-with-checklist-and-tips-optimal-cluster-usage)" width="45%" />
<p class="caption">
Figure 1.8: Image from Pradeep Redddy Raamana <a href="https://www.slideshare.net/raamana/high-performance-computing-with-checklist-and-tips-optimal-cluster-usage">High performance computing tutorial, with checklist and tips to optimize cluster usage</a>
</p>
</div>
<p>(The rest, below, is taken from the Teach Data Science blog: <a href="https://teachdatascience.com/cloud2/" class="uri">https://teachdatascience.com/cloud2/</a>, this entry written by Nick Horton)</p>
<p>The R package <code>parallel</code> is designed to send tasks to each of multiple cores. Today’s computers (even small laptops!) typically have multiple cores, and any server or cloud computing infrastructure can easily handle dozens or hundreds of parallel tasks. The structure of the R <code>parallel</code> implementation sends tasks to workers that don’t talk to one another until compiling their results at the end. In her <a href="https://rawgit.com/PPgp/useR2017public/master/tutorial.html">2017 UseR! tutorial</a>, Hana Sevcikova describes the function of workers which run code/functions/iterations separately before results are subsequently combined.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-13"></span>
<img src="figs/flow.png" alt="Image from Sevcikova UseR! 2017 [tutorial on parallel computing](https://rawgit.com/PPgp/useR2017public/master/tutorial.html)" width="65%" />
<p class="caption">
Figure 1.9: Image from Sevcikova UseR! 2017 <a href="https://rawgit.com/PPgp/useR2017public/master/tutorial.html">tutorial on parallel computing</a>
</p>
</div>
<p>As computing infrastructure becomes more sophisticated, it is important to have the language to describe how connected components work. Parallel processing allows for a conversation on the <a href="https://en.wikipedia.org/wiki/Grid_computing">differences between</a> distributed computing, cluster computing, and grid computing, and generally, the framework of high performance computing. The benefit of parallel computing as an introduction to the larger infrastructure is that the task of each worker is clear, important, and easy to describe.</p>
<p>This discussion is motivated by several recent papers and blog posts that describe how complex, real-world data science computation can be structured in ways that would not have been feasible in past years without herculean efforts. It is worth noting the <a href="https://livefreeordichotomize.com/2019/06/04/using_awk_and_r_to_parse_25tb/">fantastic example</a> that described multiple iterations needed to parse huge amounts of raw DNA sequencing data to undertake analyses for a given set of genetic locations. In <a href="https://hdsr.mitpress.mit.edu/pub/9mdn32tq">“Ambitious data science can be painless”</a> Monajemi et al. describe workflows that take advantage of new software stacks to undertake massive cloud-based experiments. While a few years older, Chamandy et al.’s <a href="https://www.tandfonline.com/doi/full/10.1080/00031305.2015.1089790">Teaching statistics at Google scale</a> described three examples where modern data challenges were overcome with creative statistical thinking (see companion report on <a href="https://ai.google/research/pubs/pub43157">Estimating uncertainty for massive data streams</a> ). Finally, the NSF-funded workshop report on <a href="https://dl.acm.org/citation.cfm?id=3233928">“Enabling computer and information science and engineering research and education in the cloud”</a> highlights opportunities as university computing migrates to cloud solutions more and more.</p>
<p>And last, you may enjoy reading the recent <a href="https://www.r-bloggers.com/three-strategies-for-working-with-big-data-in-r/">Three strategies for working with Big Data in R</a> blog post.</p>
<p>How can we prepare for cloud computing in an undergraduate course?</p>
<div id="getting-started" class="section level4 unnumbered">
<h4>Getting started</h4>
<p>What are the steps to exploring cloud-based systems? Each of the main cloud providers have active educational outreach programs.</p>
<ul>
<li><p><a href="https://edu.google.com/programs/faculty/?modal_active=none">Google Compute Platform</a> allows faculty to apply to receive $100 in GCP credits and $50 per student. Credits can be used in courses, student clubs, and other faculty-sponsored events. (To replicate our example later in this blog, you’ll want to set up an account and request credits.)</p></li>
<li><p><a href="https://azure.microsoft.com/en-us/education/">Azure for Education</a> provides access for educators to open source content for classes and $200 in Azure credits, plus free services.</p></li>
<li><p><a href="https://aws.amazon.com/education/awseducate/">Amazon Web Services Educate</a> provides between $75 and $200 in AWS annual credits per educator (depending on membership status) and between $30 and $100 for students.</p></li>
</ul>
<p>You should sign up and start to explore! The world of cloud computing is quickly changing. By gaining experience through investment in time in learning these tools will help instructors provide guidance to their students in use of these modern computational tools.</p>
</div>
<div id="an-example-bigquery-in-googles-gcp" class="section level4 unnumbered">
<h4>An example: BigQuery in Google’s GCP</h4>
<p>Consider an example using GCP (kudos to <a href="https://www.shukryzablah.com">Shukry Zablah</a> for his assistance).</p>
<p><a href="https://cloud.google.com/bigquery/">BigQuery</a> is Google’s serverless, highly-scalable, cloud data warehouse. A <a href="https://cloud.google.com/bigquery/docs/quickstarts">quickstart</a> document is available which discusses use of the web user interface and the GCP console as well as access through an API interface. The <a href="https://github.com/r-dbi/bigrquery">bigrquery</a> package in R makes it easy to work with data stored in Google BigQuery through queries to BigQuery tables.</p>
<p>The first step is to request GCP credits (see above) and use the online interface to create a project (below called “Test Project for Blog”).</p>
<div class="sourceCode" id="cb526"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb526-1" data-line-number="1"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb526-2" data-line-number="2"><span class="kw">library</span>(bigrquery)</a>
<a class="sourceLine" id="cb526-3" data-line-number="3"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb526-4" data-line-number="4"><span class="kw">library</span>(forcats)</a>
<a class="sourceLine" id="cb526-5" data-line-number="5"><span class="kw">library</span>(purrr)</a>
<a class="sourceLine" id="cb526-6" data-line-number="6"><span class="kw">library</span>(readr)</a></code></pre></div>
<div class="sourceCode" id="cb527"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb527-1" data-line-number="1">projectId &lt;-<span class="st"> &quot;bigquery-public-data&quot;</span>  <span class="co"># replace with your own project</span></a>
<a class="sourceLine" id="cb527-2" data-line-number="2">billingId &lt;-<span class="st"> &quot;test-project-for-blog&quot;</span> <span class="co"># replace with your own billing ID</span></a>
<a class="sourceLine" id="cb527-3" data-line-number="3">datasetName &lt;-<span class="st"> &quot;samples&quot;</span></a>
<a class="sourceLine" id="cb527-4" data-line-number="4">tableName &lt;-<span class="st"> &quot;wikipedia&quot;</span></a></code></pre></div>
<p>BigQuery includes a number of <a href="https://cloud.google.com/bigquery/public-data/">public datasets</a>.
Below is an analysis of the public dataset of the revisions of Wikipedia articles up to April 2010, hosted in GCP BigQuery. The size of the table is 35.69GB. The queries take only seconds to run.</p>
<div class="sourceCode" id="cb528"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb528-1" data-line-number="1">query &lt;-<span class="st"> &quot;SELECT  title, COUNT(title) as n</span></a>
<a class="sourceLine" id="cb528-2" data-line-number="2"><span class="st">          FROM `bigquery-public-data.samples.wikipedia` </span></a>
<a class="sourceLine" id="cb528-3" data-line-number="3"><span class="st">          GROUP BY title</span></a>
<a class="sourceLine" id="cb528-4" data-line-number="4"><span class="st">          ORDER BY n DESC</span></a>
<a class="sourceLine" id="cb528-5" data-line-number="5"><span class="st">          LIMIT 500&quot;</span></a></code></pre></div>
<p>For safety, always try to make sure that your queries have the <code>LIMIT</code> set on your queries.</p>
<div class="sourceCode" id="cb529"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb529-1" data-line-number="1">mostRevisions_tb &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb529-2" data-line-number="2"><span class="st">  </span>bigrquery<span class="op">::</span><span class="kw">bq_project_query</span>(<span class="dt">x =</span> billingId, </a>
<a class="sourceLine" id="cb529-3" data-line-number="3">    <span class="dt">query =</span> query) <span class="co">#creates temporary table</span></a></code></pre></div>
<p>When the previous <code>bq_project_query()</code> function is run within RStudio, a connection is made to Google (GCP) and an authentication window will open up in a local browser.</p>
<p>All the heavy lifting we perform is done on the database end (note that we are billed for it, though the first 1TB of accesses are free). The local machine only receives the data once we try to display it. Right now <code>mostRevisions_tb</code> is just a reference to a temporary table online. The query accessed 7GB of data.</p>
<p>We can get a copy of the data on our local machine once we are confident that it is what we want.</p>
<div class="sourceCode" id="cb530"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb530-1" data-line-number="1">mostRevisions &lt;-<span class="st"> </span><span class="kw">bq_table_download</span>(mostRevisions_tb) </a></code></pre></div>
<div class="sourceCode" id="cb531"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb531-1" data-line-number="1"><span class="kw">glimpse</span>(mostRevisions)</a></code></pre></div>
<pre><code>## Observations: 500
## Variables: 2
## $ title &lt;chr&gt; &quot;Wikipedia:Administrator intervention against vandalism&quot;, …
## $ n     &lt;int&gt; 643271, 419695, 326337, 257893, 226802, 204469, 191679, 18…</code></pre>
<div class="sourceCode" id="cb533"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb533-1" data-line-number="1">clean &lt;-<span class="st"> </span>mostRevisions <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb533-2" data-line-number="2"><span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&quot;Wikipedia|User|Template|Talk&quot;</span>, title)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb533-3" data-line-number="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">title =</span> <span class="kw">fct_reorder</span>(title, n)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co">#to sort levels</span></a>
<a class="sourceLine" id="cb533-4" data-line-number="4"><span class="st">  </span><span class="kw">glimpse</span>()</a></code></pre></div>
<pre><code>## Observations: 272
## Variables: 2
## $ title &lt;fct&gt; George W. Bush, List of World Wrestling Entertainment empl…
## $ n     &lt;int&gt; 43652, 30572, 27433, 23245, 21768, 20814, 20546, 20529, 20…</code></pre>
<p>Let’s plot the top 10 entries.</p>
<div class="sourceCode" id="cb535"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb535-1" data-line-number="1"><span class="kw">ggplot</span>(clean <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">10</span>), <span class="kw">aes</span>(<span class="dt">x =</span> title, <span class="dt">y =</span> n, <span class="dt">fill =</span> n)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb535-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb535-3" data-line-number="3"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Article Title&quot;</span>,</a>
<a class="sourceLine" id="cb535-4" data-line-number="4">       <span class="dt">y =</span> <span class="st">&quot;Number of Revisions&quot;</span>,</a>
<a class="sourceLine" id="cb535-5" data-line-number="5">       <span class="dt">title =</span> <span class="st">&quot;Most Revised Wikipedia Articles (Up to April 2010)&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb535-6" data-line-number="6"><span class="st">  </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;darkblue&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;darkred&quot;</span>, <span class="dt">guide =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb535-7" data-line-number="7"><span class="st">  </span><span class="kw">theme_minimal</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb535-8" data-line-number="8"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">20</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) </a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-22-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>We’ve obviously just scratched the surface here. There are lots of other examples out there to consider replicating in your classroom (e.g., <a href="https://www.mikejohnpage.com/blog/returning-tweets-on-a-schedule-in-r-using-aws-ec2-rds-and-cron/">returning tweets on a schedule</a>). Hopefully you are intrigued enough to request some credits for you and your students and start to explore. Not sure where to begin? Check out the <a href="https://www.youtube.com/playlist?list=PLIivdWyY5sqKh1gDR0WpP9iIOY00IE0xL">GCP Essentials Videos</a> series.</p>
</div>
</div>
<div id="reticulate" class="section level2">
<h2><span class="header-section-number">10.5</span> <code>reticulate</code></h2>
<p>(Taken from the Teach Data Science blog: <a href="https://teachdatascience.com/reticulate/" class="uri">https://teachdatascience.com/reticulate/</a>)</p>
<div id="connect-to-python-within-rstudio" class="section level3 unnumbered">
<h3>Connect to Python within RStudio</h3>
<p>For many statisticians, the go-to software language is R. However, there is no doubt that Python is a very important language in data science. Why not do both??</p>
<div class="sourceCode" id="cb536"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb536-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb536-2" data-line-number="2"><span class="kw">library</span>(reticulate)</a>
<a class="sourceLine" id="cb536-3" data-line-number="3"><span class="kw">use_virtualenv</span>(<span class="st">&quot;r-reticulate&quot;</span>)</a>
<a class="sourceLine" id="cb536-4" data-line-number="4">reticulate<span class="op">::</span><span class="kw">import</span>(<span class="st">&quot;statsmodels&quot;</span>)</a></code></pre></div>
<pre><code>## Module(statsmodels)</code></pre>
<div id="i-can-run-python-inside-r" class="section level4 unnumbered">
<h4>I can run Python inside R??</h4>
<p><img src="figs/pychunk1.png" width="65%" style="display: block; margin: auto auto auto 0;" />
<img src="figs/pychunk2.png" width="65%" style="display: block; margin: auto 0 auto auto;" /></p>
<ul>
<li><code>pandas</code> for data wrangling.<br />
</li>
<li>In R, the chunk is specified to be a Python chunk (RStudio is now running Python).</li>
</ul>
<pre><code>```{python}
import pandas
flights = pandas.read_csv(&quot;flights.csv&quot;)
flights = flights[flights[&quot;dest&quot;] == &quot;ORD&quot;]
flights = flights[[&#39;carrier&#39;, &#39;dep_delay&#39;, &#39;arr_delay&#39;]]
flights = flights.dropna()
```</code></pre>
<p>A view of the Python chunk which is actually run:</p>
<div class="sourceCode" id="cb539"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb539-1" data-line-number="1"><span class="im">import</span> pandas</a>
<a class="sourceLine" id="cb539-2" data-line-number="2">flights <span class="op">=</span> pandas.read_csv(<span class="st">&quot;flights.csv&quot;</span>)</a>
<a class="sourceLine" id="cb539-3" data-line-number="3">flights <span class="op">=</span> flights[flights[<span class="st">&quot;dest&quot;</span>] <span class="op">==</span> <span class="st">&quot;ORD&quot;</span>]</a>
<a class="sourceLine" id="cb539-4" data-line-number="4">flights <span class="op">=</span> flights[[<span class="st">&#39;carrier&#39;</span>, <span class="st">&#39;dep_delay&#39;</span>, <span class="st">&#39;arr_delay&#39;</span>]]</a>
<a class="sourceLine" id="cb539-5" data-line-number="5">flights <span class="op">=</span> flights.dropna()</a></code></pre></div>
</div>
<div id="learn-about-the-dataset" class="section level4 unnumbered">
<h4>Learn about the dataset</h4>
<pre><code>```{python}
flights.shape
flights.head(3)
flights.describe()
```</code></pre>
<div class="sourceCode" id="cb541"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb541-1" data-line-number="1">flights.shape</a></code></pre></div>
<pre><code>## (12590, 3)</code></pre>
<div class="sourceCode" id="cb543"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb543-1" data-line-number="1">flights.head(<span class="dv">3</span>)</a></code></pre></div>
<pre><code>##    carrier  dep_delay  arr_delay
## 4       UA       -4.0       12.0
## 5       AA       -2.0        8.0
## 22      AA       -1.0       14.0</code></pre>
<div class="sourceCode" id="cb545"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb545-1" data-line-number="1">flights.describe()</a></code></pre></div>
<pre><code>##           dep_delay     arr_delay
## count  12590.000000  12590.000000
## mean      11.709770      2.917951
## std       39.409704     44.885155
## min      -20.000000    -62.000000
## 25%       -6.000000    -22.000000
## 50%       -2.000000    -10.000000
## 75%        9.000000     10.000000
## max      466.000000    448.000000</code></pre>
</div>
<div id="computations-using-pandas" class="section level4 unnumbered">
<h4>Computations using <code>pandas</code></h4>
<pre><code>```{python}
flights = pandas.read_csv(&quot;flights.csv&quot;)
flights = flights[[&#39;carrier&#39;, &#39;dep_delay&#39;, &#39;arr_delay&#39;]]
flights.groupby(&quot;carrier&quot;).mean()
```</code></pre>
<div class="sourceCode" id="cb548"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb548-1" data-line-number="1">flights <span class="op">=</span> pandas.read_csv(<span class="st">&quot;flights.csv&quot;</span>)</a>
<a class="sourceLine" id="cb548-2" data-line-number="2">flights <span class="op">=</span> flights[[<span class="st">&#39;carrier&#39;</span>, <span class="st">&#39;dep_delay&#39;</span>, <span class="st">&#39;arr_delay&#39;</span>]]</a>
<a class="sourceLine" id="cb548-3" data-line-number="3">flights.groupby(<span class="st">&quot;carrier&quot;</span>).mean()</a></code></pre></div>
<pre><code>##          dep_delay  arr_delay
## carrier                      
## AA        8.586016   0.364291
## AS        5.804775  -9.930889
## DL        9.264505   1.644341
## UA       12.106073   3.558011
## US        3.782418   2.129595</code></pre>
</div>
<div id="from-python-chunk-to-r-chunk" class="section level4 unnumbered">
<h4>From Python chunk to R chunk</h4>
<ul>
<li><code>py$x</code> accesses an <code>x</code> variable created within Python from R</li>
<li><code>r.x</code> accesses an <code>x</code> variable created within R from Python</li>
</ul>
<div class="sourceCode" id="cb550"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb550-1" data-line-number="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb550-2" data-line-number="2"><span class="kw">ggplot</span>(py<span class="op">$</span>flights, </a>
<a class="sourceLine" id="cb550-3" data-line-number="3">       <span class="kw">aes</span>(<span class="dt">x=</span>carrier, </a>
<a class="sourceLine" id="cb550-4" data-line-number="4">           <span class="dt">y=</span>arr_delay)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb550-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb550-6" data-line-number="6"><span class="st">  </span><span class="kw">geom_jitter</span>()</a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-30-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="from-r-chunk-to-python-chunk" class="section level4 unnumbered">
<h4>From R chunk to Python chunk</h4>
<div class="sourceCode" id="cb551"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb551-1" data-line-number="1"><span class="kw">data</span>(diamonds)</a>
<a class="sourceLine" id="cb551-2" data-line-number="2"><span class="kw">head</span>(diamonds)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 10
##   carat cut       color clarity depth table price     x     y     z
##   &lt;dbl&gt; &lt;ord&gt;     &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 0.23  Ideal     E     SI2      61.5    55   326  3.95  3.98  2.43
## 2 0.21  Premium   E     SI1      59.8    61   326  3.89  3.84  2.31
## 3 0.23  Good      E     VS1      56.9    65   327  4.05  4.07  2.31
## 4 0.290 Premium   I     VS2      62.4    58   334  4.2   4.23  2.63
## 5 0.31  Good      J     SI2      63.3    58   335  4.34  4.35  2.75
## 6 0.24  Very Good J     VVS2     62.8    57   336  3.94  3.96  2.48</code></pre>
</div>
<div id="python-chunks" class="section level4 unnumbered">
<h4>Python chunks</h4>
<p>Note that we’re calling Python code on an R object.</p>
<div class="sourceCode" id="cb553"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb553-1" data-line-number="1"><span class="bu">print</span>(r.diamonds.describe())</a></code></pre></div>
<pre><code>##               carat         depth  ...             y             z
## count  53940.000000  53940.000000  ...  53940.000000  53940.000000
## mean       0.797940     61.749405  ...      5.734526      3.538734
## std        0.474011      1.432621  ...      1.142135      0.705699
## min        0.200000     43.000000  ...      0.000000      0.000000
## 25%        0.400000     61.000000  ...      4.720000      2.910000
## 50%        0.700000     61.800000  ...      5.710000      3.530000
## 75%        1.040000     62.500000  ...      6.540000      4.040000
## max        5.010000     79.000000  ...     58.900000     31.800000
## 
## [8 rows x 7 columns]</code></pre>
<div class="sourceCode" id="cb555"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb555-1" data-line-number="1"><span class="im">import</span> statsmodels.formula.api <span class="im">as</span> smf</a>
<a class="sourceLine" id="cb555-2" data-line-number="2">model <span class="op">=</span> smf.ols(<span class="st">&#39;price ~ carat&#39;</span>, data <span class="op">=</span> r.diamonds).fit()</a>
<a class="sourceLine" id="cb555-3" data-line-number="3"><span class="bu">print</span>(model.summary())</a></code></pre></div>
<pre><code>##                             OLS Regression Results                            
## ==============================================================================
## Dep. Variable:                  price   R-squared:                       0.849
## Model:                            OLS   Adj. R-squared:                  0.849
## Method:                 Least Squares   F-statistic:                 3.041e+05
## Date:                Sun, 19 Jul 2020   Prob (F-statistic):               0.00
## Time:                        08:22:16   Log-Likelihood:            -4.7273e+05
## No. Observations:               53940   AIC:                         9.455e+05
## Df Residuals:                   53938   BIC:                         9.455e+05
## Df Model:                           1                                         
## Covariance Type:            nonrobust                                         
## ==============================================================================
##                  coef    std err          t      P&gt;|t|      [0.025      0.975]
## ------------------------------------------------------------------------------
## Intercept  -2256.3606     13.055   -172.830      0.000   -2281.949   -2230.772
## carat       7756.4256     14.067    551.408      0.000    7728.855    7783.996
## ==============================================================================
## Omnibus:                    14025.341   Durbin-Watson:                   0.986
## Prob(Omnibus):                  0.000   Jarque-Bera (JB):           153030.525
## Skew:                           0.939   Prob(JB):                         0.00
## Kurtosis:                      11.035   Cond. No.                         3.65
## ==============================================================================
## 
## Warnings:
## [1] Standard Errors assume that the covariance matrix of the errors is correctly specified.</code></pre>
</div>
<div id="running-just-python" class="section level4 unnumbered">
<h4>Running just Python</h4>
<p><img src="figs/PyScript.png" width="120%" style="display: block; margin: auto;" /></p>
</div>
<div id="full-disclosure" class="section level4 unnumbered">
<h4>Full disclosure</h4>
<p><code>reticulate</code> is not always trivial to set up. Indeed, I’ve had trouble figuring out which Python version is talking to R and where different module versions live.</p>
</div>
<div id="learn-more-1" class="section level4 unnumbered">
<h4>Learn more</h4>
<ul>
<li><a href="https://rstudio.github.io/reticulate/">RStudio R Interface to Python</a></li>
</ul>
<p><a href="https://rstudio.github.io/reticulate/" class="uri">https://rstudio.github.io/reticulate/</a></p>
<ul>
<li><a href="https://blog.rstudio.com/2018/10/09/rstudio-1-2-preview-reticulated-python/">RStudio blog on Reticulated Python</a></li>
</ul>
<p><a href="https://blog.rstudio.com/2018/10/09/rstudio-1-2-preview-reticulated-python" class="uri">https://blog.rstudio.com/2018/10/09/rstudio-1-2-preview-reticulated-python</a></p>
</div>
</div>
</div>
<div id="sql-in-r" class="section level2">
<h2><span class="header-section-number">10.6</span> SQL (in R)</h2>
<p>Note that there exists an R interface to work with SQL commands from within an R Markdown file. For consistency with the class notes, we’ve continued to use the R Markdown structure to demonstrate the course material.</p>
<p>(Taken from the Teach Data Science blog: <a href="https://teachdatascience.com/sql/" class="uri">https://teachdatascience.com/sql/</a>, this entry written by Nick Horton)</p>
<p>SQL (pronounced <em>sequel</em>) stands for <a href="https://en.wikipedia.org/wiki/SQL">Structured Query Language</a>; it is a language designed to manage data in a relational database system.</p>
<p>We will use a public facing MySQL <a href="http://www.science.smith.edu/wai-database/home/about">database</a> containing wideband acoustic immittance (WAI) measures made on normal ears of adults. (The project is funded by the National Institutes of Health, NIDCD, and hosted on a server at Smith College, <a href="https://projectreporter.nih.gov/project_info_description.cfm?aid=8769352&amp;icde=44962011&amp;ddparam=&amp;ddvalue=&amp;ddsub=&amp;cr=46&amp;csb=default&amp;cs=ASC&amp;pball=">PI Susan Voss, R15 DC014129-01</a>.) The database was created to enable auditory researchers to share WAI measurements and combine analyses over multiple datasets.</p>
<p>We begin by demonstrating how SQL queries can be sent to a database. It is necessary to set up a connection using the <code>dbConnect()</code> function.</p>
<div class="sourceCode" id="cb557"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb557-1" data-line-number="1"><span class="kw">library</span>(mosaic)</a>
<a class="sourceLine" id="cb557-2" data-line-number="2"><span class="kw">library</span>(RMySQL)  </a>
<a class="sourceLine" id="cb557-3" data-line-number="3">con &lt;-<span class="st"> </span><span class="kw">dbConnect</span>(</a>
<a class="sourceLine" id="cb557-4" data-line-number="4">  RMySQL<span class="op">::</span><span class="kw">MySQL</span>(), <span class="dt">host =</span> <span class="st">&quot;scidb.smith.edu&quot;</span>, <span class="dt">user =</span> <span class="st">&quot;waiuser&quot;</span>, </a>
<a class="sourceLine" id="cb557-5" data-line-number="5">  <span class="dt">password =</span> <span class="st">&quot;smith_waiDB&quot;</span>, <span class="dt">dbname =</span> <span class="st">&quot;wai&quot;</span>)</a></code></pre></div>
<p>Next a series of SQL queries can be sent to the database using the <code>DBI::dbGetQuery()</code> function: each query returns an R dataframe.</p>
<div class="sourceCode" id="cb558"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb558-1" data-line-number="1"><span class="kw">class</span>(<span class="kw">dbGetQuery</span>(con, <span class="st">&quot;SHOW TABLES&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>There are multiple tables within the <code>wai</code> database.</p>
<div class="sourceCode" id="cb560"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb560-1" data-line-number="1"><span class="kw">dbGetQuery</span>(con, <span class="st">&quot;SHOW TABLES&quot;</span>)</a></code></pre></div>
<pre><code>##       Tables_in_wai
## 1          Codebook
## 2      Measurements
## 3 Measurements_2020
## 4           PI_Info
## 5       PI_Info_OLD
## 6           Subject
## 7     Subjects_2020</code></pre>
<p>The <code>EXPLAIN</code> command describes the ten field names (variables) in the <code>PI_Info</code> table.</p>
<div class="sourceCode" id="cb562"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb562-1" data-line-number="1"><span class="kw">dbGetQuery</span>(con, <span class="st">&quot;EXPLAIN PI_Info&quot;</span>)</a></code></pre></div>
<pre><code>##                 Field        Type Null Key Default Extra
## 1          Identifier varchar(50)   NO PRI    &lt;NA&gt;      
## 2                Year     int(11)   NO        &lt;NA&gt;      
## 3             Authors        text   NO        &lt;NA&gt;      
## 4    AuthorsShortList        text   NO        &lt;NA&gt;      
## 5               Title        text   NO        &lt;NA&gt;      
## 6             Journal        text   NO        &lt;NA&gt;      
## 7                 URL        text   NO        &lt;NA&gt;      
## 8            Abstract        text   NO        &lt;NA&gt;      
## 9   DataSubmitterName        text   NO        &lt;NA&gt;      
## 10 DataSubmitterEmail        text   NO        &lt;NA&gt;      
## 11      DateSubmitted        text   NO        &lt;NA&gt;      
## 12           PI_Notes        text   NO        &lt;NA&gt;</code></pre>
<p>The <code>SELECT</code> statement can be used to select all fields for eight observations in the <code>Measurements</code> table.</p>
<div class="sourceCode" id="cb564"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb564-1" data-line-number="1">eightobs &lt;-<span class="st"> </span><span class="kw">dbGetQuery</span>(con, <span class="st">&quot;SELECT * FROM Measurements LIMIT 8&quot;</span>)</a>
<a class="sourceLine" id="cb564-2" data-line-number="2">eightobs</a></code></pre></div>
<pre><code>##   Identifier Sub_Number Session Left_Ear MEP Instrument    Freq Absorbance
## 1  Abur_2014          1       1        0  -5          1 210.938  0.0451375
## 2  Abur_2014          1       1        0  -5          1 234.375  0.0441247
## 3  Abur_2014          1       1        0  -5          1 257.812  0.0495935
## 4  Abur_2014          1       1        0  -5          1 281.250  0.0516088
## 5  Abur_2014          1       1        0  -5          1 304.688  0.0590836
## 6  Abur_2014          1       1        0  -5          1 328.125  0.0628038
## 7  Abur_2014          1       1        0  -5          1 351.562  0.0682962
## 8  Abur_2014          1       1        0  -5          1 375.000  0.0738373
##        Zmag      Zang Canal_Area
## 1 110638000 -0.228113         NA
## 2 100482000 -0.230561         NA
## 3  90561100 -0.230213         NA
## 4  83515500 -0.230959         NA
## 5  77476800 -0.229652         NA
## 6  71229100 -0.230026         NA
## 7  66615500 -0.229576         NA
## 8  61996200 -0.229327         NA</code></pre>
<p>More interesting and complicated <code>SELECT</code> calls can be used to undertake grouping and aggregation. Here we calculate the sample size for each study</p>
<div class="sourceCode" id="cb566"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb566-1" data-line-number="1"><span class="kw">dbGetQuery</span>(con, </a>
<a class="sourceLine" id="cb566-2" data-line-number="2">  <span class="st">&quot;SELECT Identifier, count(*) AS NUM FROM Measurements GROUP BY Identifier ORDER BY NUM&quot;</span>)</a></code></pre></div>
<pre><code>##       Identifier    NUM
## 1       Sun_2016   2604
## 2    Shaver_2013   2880
## 3    Feeney_2017   3162
## 4      Voss_1994   5120
## 5       Liu_2008   5520
## 6    Werner_2010   7935
## 7  Rosowski_2012  14384
## 8      Voss_2010  14880
## 9      Abur_2014  21328
## 10    Groon_2015  35469
## 11  Shahnaz_2006  58776
## 12    Lewis_2015 114716</code></pre>
<div id="accessing-a-database-using-dplyr-commands" class="section level4 unnumbered">
<h4>Accessing a database using <code>dplyr</code> commands</h4>
<p>Alternatively, a connection can be made to the server by creating a series of <code>dplyr</code> tbl
objects. Connecting with familiar <code>dplyr</code> syntax is attractive because, as <a href="https://dbplyr.tidyverse.org/articles/dbplyr.html">Hadley Wickham</a> has noted, SQL and R have similar syntax (but sufficiently different to be confusing).</p>
<p>The setup process looks similar.</p>
<div class="sourceCode" id="cb568"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb568-1" data-line-number="1">Measurements &lt;-<span class="st"> </span><span class="kw">tbl</span>(con, <span class="st">&quot;Measurements&quot;</span>)</a>
<a class="sourceLine" id="cb568-2" data-line-number="2"><span class="kw">class</span>(Measurements)</a></code></pre></div>
<pre><code>## [1] &quot;tbl_MySQLConnection&quot; &quot;tbl_dbi&quot;             &quot;tbl_sql&quot;            
## [4] &quot;tbl_lazy&quot;            &quot;tbl&quot;</code></pre>
<div class="sourceCode" id="cb570"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb570-1" data-line-number="1">PI_Info &lt;-<span class="st"> </span><span class="kw">tbl</span>(con, <span class="st">&quot;PI_Info&quot;</span>)</a>
<a class="sourceLine" id="cb570-2" data-line-number="2">Subject &lt;-<span class="st"> </span><span class="kw">tbl</span>(con, <span class="st">&quot;Subject&quot;</span>)</a></code></pre></div>
<p>We explore the <code>PI_Info</code> table using the <code>collect()</code> function used to force computation on the database (and return the results). One attractive aspect of database systems is that they feature <a href="https://en.wikipedia.org/wiki/Lazy_evaluation"><em>lazy evaluation</em></a>, where computation is optimized and postponed as long as possible.</p>
<div class="sourceCode" id="cb571"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb571-1" data-line-number="1">PI_Info  <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">total =</span> <span class="kw">n</span>())</a></code></pre></div>
<pre><code>## # Source:   lazy query [?? x 1]
## # Database: mysql 5.5.58-0ubuntu0.14.04.1-log [waiuser@scidb.smith.edu:/wai]
##   total
##   &lt;dbl&gt;
## 1    12</code></pre>
<div class="sourceCode" id="cb573"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb573-1" data-line-number="1">PI_Info <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">collect</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">data.frame</span>()   </a></code></pre></div>
<pre><code>##       Identifier Year
## 1      Abur_2014 2014
## 2    Feeney_2017 2017
## 3     Groon_2015 2015
## 4     Lewis_2015 2015
## 5       Liu_2008 2008
## 6  Rosowski_2012 2012
## 7   Shahnaz_2006 2006
## 8    Shaver_2013 2013
## 9       Sun_2016 2016
## 10     Voss_1994 1994
## 11     Voss_2010 2010
## 12   Werner_2010 2010
##                                                                                                                                          Authors
## 1                                                                                              Defne Abur, Nicholas J. Horton, and Susan E. Voss
## 2   M. Patrick Feeney,  Douglas H. Keefe, Lisa L. Hunter,  Denis F. Fitzpatrick, Angela C. Garinis, Daniel B. Putterman, and Garnett P. McMillan
## 3                                               Katherine A. Groon, Daniel M. Rasetshwane, Judy G. Kopun, Michael P. Gorga, and Stephen T. Neely
## 4                                                                                                            James D. Lewis and Stephen T. Neely
## 5                                  Yi-Wen Liu, Chris A. Sanford,  John C. Ellison, Denis F. Fitzpatrick,  Michael P. Gorga, and Douglas H. Keefe
## 6  John J. Rosowski, Hideko H. Nakajima, Mohamad A. Hamade, Lorice Mahfoud, Gabrielle R. Merchant, Christopher F. Halpin, and Saumil N. Merchant
## 7                                                                                                                   Navid Shahnaz and Karin Bork
## 8                                                                                                               Mark D. Shaver and Xiao-Ming Sun
## 9                                                                                                                                  Xiao-Ming Sun
## 10                                                                                                               Susan E. Voss and Jont B. Allen
## 11                               Susan E. Voss, Modupe F. Adegoke, Nicholas J. Horton, Kevin N. Sheth, Jonathan Rosand, and Christopher A. Shera
## 12                                                                                          Lynne A. Werner, Ellen C. Levi, and Douglas H. Keefe
##    AuthorsShortList
## 1       Abur et al.
## 2     Feeney et al.
## 3      Groon et al.
## 4   Lewis and Neely
## 5        Liu et al.
## 6   Rosowski et al.
## 7  Shahnaz and Bork
## 8    Shaver and Sun
## 9               Sun
## 10   Voss and Allen
## 11      Voss et al.
## 12    Werner et al.
##                                                                                                                                     Title
## 1                                                                                          Instrasubject variability in power reflectance
## 2       Normative wideband reflectance, equivalent admittance at the tympanic membrane, and acoustic stapedius reflex threshold in adults
## 3                                                                                       Air-leak effects on ear-canal acoustic absorbance
## 4                                                                    Non-invasive estimation of middle-ear input impedance and efficiency
## 5                    Wideband absorbance tympanometry using pressure sweeps: System development and results on adults with normal hearing
## 6                                                         Ear-canal reflectance, umbo velocity, and tympanometry in normal-hearing adults
## 7                                                                       Wideband reflectance norms for caucasian and chinese young adults
## 8  Wideband energy reflectance measurements: Effects of negative middle ear pressure and application of a pressure compensation procedure
## 9                       Wideband acoustic immittance: Normative study and test-retest reliability of tympanometric measurements in adults
## 10                                                               Measurement of acoustic impedance and reflectance in the human ear canal
## 11                                                               Posture systematically alters ear-canal reflectance and DPOAE properties
## 12                                            Ear-canal wideband acoustic transfer functions of adults and two- to nine-month-old infants
##                   Journal
## 1        J Am Acad Audiol
## 2                Ear Hear
## 3                Ear Hear
## 4         J Acoust Soc Am
## 5         J Acoust Soc Am
## 6                Ear Hear
## 7                Ear Hear
## 8         J Acoust Soc Am
## 9  J Speech Lang Hear Res
## 10        J Acoust Soc Am
## 11               Hear Res
## 12               Ear Hear
##                                                                                                                   URL
## 1                                                                        https://www.ncbi.nlm.nih.gov/pubmed/25257718
## 2                                                                        https://www.ncbi.nlm.nih.gov/pubmed/28045835
## 3  https://journals.lww.com/ear-hearing/fulltext/2015/01000/Air_Leak_Effects_on_Ear_Canal_Acoustic_Absorbance.16.aspx
## 4                                                                 https://asa.scitation.org/doi/abs/10.1121/1.4927408
## 5                                                                        https://www.ncbi.nlm.nih.gov/pubmed/19206798
## 6                                                                         http://www.ncbi.nlm.nih.gov/pubmed/21857517
## 7        http://journals.lww.com/ear-hearing/Abstract/2006/12000/Wideband_Reflectance_Norms_for_Caucasian_and.15.aspx
## 8                                                            &quot;\nhttps://asa.scitation.org/doi/full/10.1121/1.4807509&quot;
## 9                                                                        https://www.ncbi.nlm.nih.gov/pubmed/27517667
## 10                                                                 https://asa.scitation.org/doi/abs/10.1121/1.408329
## 11                                                                       https://www.ncbi.nlm.nih.gov/pubmed/20227475
## 12                                                                       https://www.ncbi.nlm.nih.gov/pubmed/20517155
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Abstract
## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         &quot;&lt;p&gt; &lt;strong&gt; Background: &lt;/strong&gt; Power reflectance measurements are an active area of research related to the development of noninvasive middle-ear assessment methods. There are limited data related to test-retest measures of power reflectance. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Purpose: &lt;/strong&gt; This study investigates test-retest features of power reflectance, including comparisons of intrasubject versus intersubject variability and how ear-canal measurement location affects measurements. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Research design: &lt;/strong&gt; Repeated measurements of power reflectance were made at about weekly intervals. The subjects returned for four to eight sessions. Measurements were made at three ear-canal locations: a deep insertion depth (with a foam plug flush at the entrance to the ear canal) and both 3 and 6 mm more lateral to this deep insertion. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Study sample:&lt;/strong&gt; Repeated measurements on seven subjects are reported. All subjects were female, between 19 and 22 yr old, and enrolled at an undergraduate women&#39;s college. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Data collection and analysis: &lt;/strong&gt; Measurements on both the right and left ears were made at three ear-canal locations during each of four to eight measurement sessions. Random-effects regression models were used for the analysis to account for repeated measures within subjects. The mean power reflectance for each position over all sessions was calculated for each subject. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Results: &lt;/strong&gt; The comparison of power reflectance from the left and right ears of an individual subject varied greatly over the seven subjects; the difference between the power reflectance measured on the left and that measured on the right was compared at 248 frequencies, and depending on the subject, the percentage of tested frequencies for which the left and right ears differed significantly ranged from 10% to 93% (some with left values greater than right values and others with the opposite pattern). Although the individual subjects showed left-right differences, the overall population generally did not show significant differences between the left and right ears. The mean power reflectance for each measurement position over all sessions depended on the location of the probe in the ear for frequencies of less than 1000 Hz. The standard deviation between subjects&#39; mean power reflectance after controlling for ear (left or right) was found to be greater than the standard deviation within the individual subject&#39;s mean power reflectance. The intrasubject standard deviation in power reflectance was smallest at the deepest insertion depths. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Conclusions: &lt;/strong&gt;  All subjects had differences in power reflectance between their left and right ears at some frequencies; the percentage of frequencies at which differences occurred varied greatly across subjects. The intrasubject standard deviations were smallest for the deepest probe insertion depths, suggesting clinical measurements should be made with as deep an insertion as practically possible to minimize variability. This deep insertion will reduce both acoustic leaks and the effect of low-frequency ear-canal losses. The within-subject standard deviations were about half the magnitude of the overall standard deviations, quantifying the extent of intrasubject versus intersubject variability. &lt;/p&gt;&quot;
## 2  &quot;&lt;p&gt; &lt;strong&gt; Objectives: &lt;/strong&gt; Wideband acoustic immittance (WAI) measures such as\npressure reflectance, parameterized by absorbance and group delay,\nequivalent admittance at the tympanic membrane (TM), and acoustic\nstapedius reflex threshold (ASRT) describe middle ear function across a\nwide frequency range, compared with traditional tests employing a single\nfrequency. The objective of this study was to obtain normative data\nusing these tests for a group of normal-hearing adults and investigate\ntest-retest reliability using a longitudinal design. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Design: &lt;/strong&gt; A longitudinal prospective design was used to obtain normative\ntest and retest data on clinical and WAI measures. Subjects were 13\nmales and 20 females (mean age = 26 years). Inclusion criteria included\nnormal audiometry and clinical immittance. Subjects were tested on two\nseparate visits approximately 1 month apart. Reflectance and equivalent\nadmittance at the TM were measured from 0.25 to 8.0 kHz under\nthree conditions: at ambient pressure in the ear canal and with pressure\nsweeps from positive to negative pressure (downswept) and negative to\npositive pressure (upswept). Equivalent admittance at the TM was calculated\nusing admittance measurements at the probe tip that were adjusted\nusing a model of sound transmission in the ear canal and acoustic estimates\nof ear-canal area and length. Wideband ASRTs were measured\nat tympanometric peak pressure (TPP) derived from the average TPP\nof downswept and upswept tympanograms. Descriptive statistics were\nobtained for all WAI responses, and wideband and clinical ASRTs were\ncompared. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Results: &lt;/strong&gt; Mean absorbance at ambient pressure and TPP demonstrated\na broad band-pass pattern typical of previous studies. Test-retest differences\nwere lower for absorbance at TPP for the downswept method\ncompared with ambient pressure at frequencies between 1.0 and\n1.26 kHz. Mean tympanometric peak-to-tail differences for absorbance\nwere greatest around 1.0 to 2.0 kHz and similar for positive and negative\ntails. Mean group delay at ambient pressure and at TPP were greatest\nbetween 0.32 and 0.6 kHz at 200 to 300 &amp;mu;sec, reduced at frequencies\nbetween 0.8 and 1.5 kHz, and increased above 1.5 kHz to around 150\n&amp;mu;sec. Mean equivalent admittance at the TM had a lower level for the\nambient method than at TPP for both sweep directions below 1.2 kHz,\nbut the difference between methods was only statistically significant for\nthe comparison between the ambient method and TPP for the upswept\ntympanogram. Mean equivalent admittance phase was positive at all frequencies.\nTest-retest reliability of the equivalent admittance level ranged\nfrom 1 to 3 dB at frequencies below 1.0 kHz, but increased to 8 to 9\ndB at higher frequencies. The mean wideband ASRT for an ipsilateral\nbroadband noise activator was 12 dB lower than the clinical ASRT, but\nhad poorer reliability. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Conclusions: &lt;/strong&gt;  Normative data for the WAI test battery revealed minor\ndifferences for results at ambient pressure compared with tympanometric methods at TPP for reflectance, group delay, and equivalent\nadmittance level at the TM for subjects with middle ear pressure within\n&amp;plusmn;100 daPa. Test-retest reliability was better for absorbance at TPP for\nthe downswept tympanogram compared with ambient pressure at frequencies\naround 1.0 kHz. Large peak-to-tail differences in absorbance\ncombined with good reliability at frequencies between about 0.7 and\n3.0 kHz suggest that this may be a sensitive frequency range for interpreting\nabsorbance at TPP. The mean wideband ipsilateral ASRT was\nlower than the clinical ASRT, consistent with previous studies. Results\nare promising for the use of a wideband test battery to evaluate middle\near function. &lt;/p&gt;\n&lt;p&gt; &lt;strong&gt; Key words: &lt;/strong&gt; Absorbed sound power, Acoustic stapedius reflex threshold,\nReflectance, Test-retest reliability, Wideband acoustic immittance. &lt;/p&gt;&quot;
## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   &quot;&lt;p&gt; &lt;strong&gt; Objective: &lt;/strong&gt; Accurate ear-canal acoustic measurements, such as wideband\nacoustic admittance, absorbance, and otoacoustic emissions,\nrequire that the measurement probe be tightly sealed in the ear canal.\nAir leaks can compromise the validity of the measurements, interfere\nwith calibrations, and increase variability. There are no established procedures\nfor determining the presence of air leaks or criteria for what size\nleak would affect the accuracy of ear-canal acoustic measurements. The\npurpose of this study was to determine ways to quantify the effects of air\nleaks and to develop objective criteria to detect their presence. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Design: &lt;/strong&gt; Air leaks were simulated by modifying the foam tips that are\nused with the measurement probe through insertion of thin plastic\ntubing. To analyze the effect of air leaks, acoustic measurements were\ntaken with both modified and unmodified foam tips in brass-tube cavities\nand human ear canals. Measurements were initially made in cavities\nto determine the range of critical leaks. Subsequently, data were collected\nin ears of 21 adults with normal hearing and normal middle-ear\nfunction. Four acoustic metrics were used for predicting the presence of\nair leaks and for quantifying these leaks: (1) low-frequency admittance\nphase (averaged over 0.1-0.2 kHz), (2) low-frequency absorbance, (3)\nthe ratio of compliance volume to physical volume (CV/PV), and (4) the\nair-leak resonance frequency. The outcome variable in this analysis was\nthe absorbance change (&amp;Delta;absorbance), which was calculated in eight\nfrequency bands. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Results: &lt;/strong&gt; The trends were similar for both the brass cavities and the ear\ncanals. &amp;Delta;Absorbance generally increased with air-leak size and was largest\nfor the lower frequency bands (0.1-0.2 and 0.2-0.5 kHz). Air-leak\neffects were observed in frequencies up to 10 kHz, but their effects above\n1 kHz were unpredictable. These high-frequency air leaks were larger in\nbrass cavities than in ear canals. Each of the four predictor variables\nexhibited consistent dependence on air-leak size. Low-frequency admittance\nphase and CV/PV decreased, while low-frequency absorbance and\nthe air-leak resonance frequency increased. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Conclusion: &lt;/strong&gt;  The effect of air leaks can be significant when their equivalent\ndiameter exceeds 0.01 in. The observed effects were greatest at low frequencies\nwhere air leaks caused absorbance to increase. Recommended\ncriteria for detecting air leaks include the following: when the frequency\nrange of interest extends as low as 0.1 kHz, low-frequency absorbance\nshould be &amp;#8804;0.20 and low-frequency admittance phase &amp;#8805;61 degrees. For\nfrequency ranges as low as 0.2 kHz, low-frequency absorbance should\nbe &amp;#8804;0.29 and low-frequency admittance phase &amp;#8805;44 degrees. &lt;/p&gt;&quot;
## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    &quot;&lt;p&gt; A method to transform the impedance measured in the ear canal, &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt;, to the plane of the eardrum,\n&lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt;, is described. The portion of the canal between the probe and eardrum was modeled as a concatenated\nseries of conical segments, allowing for spatial variations in its cross-sectional area. A\nmodel of the middle ear (ME) and cochlea terminated the ear-canal model, which permitted estimation\nof ME efficiency. Acoustic measurements of &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt; were made at two probe locations in 15\nnormal-hearing subjects. &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt; was sensitive to measurement location, especially near frequencies of\ncanal resonances and anti-resonances. Transforming &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt; to &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt; reduced the influence of the canal,\ndecreasing insertion-depth sensitivity of &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt; between 1 and 12 kHz compared to &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt;. Absorbance,\nA, was less sensitive to probe placement than &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt;, but more sensitive than &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt; above 5 kHz. &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt;\nand A were similarly insensitive to probe placement between 1 and 5 kHz. The probe-placement\nsensitivity of &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt; below 1 kHz was not reduced from that of either A or &lt;i&gt;Z&lt;sub&gt;EC&lt;/sub&gt;&lt;/i&gt;. ME efficiency had a\nbandpass shape with greatest efficiency between 1 and 4 kHz. Estimates of &lt;i&gt;Z&lt;sub&gt;ED&lt;/sub&gt;&lt;/i&gt; and ME efficiency\ncould extend the diagnostic capability of wideband-acoustic immittance measurements.&lt;/p&gt;&quot;
## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                &quot;&lt;p&gt; A system with potential for middle-ear screening and diagnostic testing was developed for the\nmeasurement of wideband energy absorbance (EA) in the ear canal as a function of air pressure, and\ntested on adults with normal hearing. Using a click stimulus, the EA was measured at 60 frequencies\nbetween 0.226 and 8 kHz. Ambient-pressure results were similar to past studies. To perform\ntympanometry, air pressure in the ear canal was controlled automatically to sweep between &amp;#x2212;300\nand 200 daPa (ascending/descending directions) using sweep speeds of approximately 75, 100, 200,\nand 400 daPa/ s. Thus, the measurement time for wideband tympanometry ranged from 1.5 to 7 s\nand was suitable for clinical applications. A bandpass tympanogram, calculated for each ear by\nfrequency averaging EA from 0.38 to 2 kHz, had a single-peak shape; however, its tympanometric\npeak pressure (TPP) shifted as a function of sweep speed and direction. EA estimated at the TPP was\nsimilar across different sweep speeds, but was higher below 2 kHz than EA measured at ambient\npressure. Future studies of EA on normal ears of a different age group or on impaired ears may be\ncompared with the adult normal baseline obtained in this study.&lt;/p&gt;&quot;
## 6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              &quot;&lt;p&gt; &lt;strong&gt; Objective: &lt;/strong&gt; This study compares measurements of ear-canal reflectance\n(ECR) to other objective measurements of middle ear function including\naudiometry, umbo velocity (&lt;i&gt;V&lt;sub&gt;U&lt;/sub&gt;&lt;/i&gt;), and tympanometry in a population of\nstrictly defined normal-hearing ears. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Design: &lt;/strong&gt; Data were prospectively gathered from 58 ears of 29 normalhearing\nsubjects, 16 females and 13 males, aged 22 to 64 yr. Subjects\nmet all of the following criteria to be considered as having normal\nhearing: (1) no history of significant middle ear disease; (2) no history\nof otologic surgery; (3) normal tympanic membrane on otoscopy; (4)\npure-tone audiometric thresholds of 20 dB HL or better for 0.25 to 8\nkHz; (5) air-bone gaps no greater than 15 dB at 0.25 kHz and 10 dB for\n0.5 to 4 kHz; (6) normal, type-A peaked tympanograms; and (7) all\nsubjects had two &quot;&quot;normal&quot;&quot; ears (as defined by these criteria). Measurements\nincluded pure-tone audiometry for 0.25 to 8 kHz, standard\n226 Hz tympanometry, ECR for 0.2 to 6 kHz at 60 dB SPL using the\nMimosa Acoustics HearID system, and umbo velocity (&lt;i&gt;V&lt;sub&gt;U&lt;/sub&gt;&lt;/i&gt;) for 0.3 to\n6 kHz at 70 to 90 dB SPL using the HLV-1000 laser Doppler\nvibrometer (Polytec Inc). &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Results: &lt;/strong&gt; Mean power reflectance (|ECR|&lt;sup&gt;2&lt;/sup&gt;) was near 1.0 at 0.2 to 0.3 kHz,\ndecreased to a broad minimum of 0.3 to 0.4 between 1 and 4 kHz, and\nthen sharply increased to almost 0.8 by 6 kHz. The mean pressure\nreflectance phase angle (&amp;#8736;ECR) plotted on a linear frequency scale\nshowed a group delay of approximately 0.1 msec for 0.2 to 6 kHz. Small\nsignificant differences were observed in |ECR|&lt;sup&gt;2&lt;/sup&gt; at the lowest frequencies\nbetween right and left ears and between males and females at 4 kHz.\n|ECR|&lt;sup&gt;2&lt;/sup&gt; decreased with age but reached significance only at 1 kHz. Our\nECR measurements were generally similar to previous published reports.\nHighly significant negative correlations were found between\n|ECR|&lt;sup&gt;2&lt;/sup&gt; and &lt;i&gt;V&lt;sub&gt;U&lt;/sub&gt;&lt;/i&gt; for frequencies below 1 kHz. Significant correlations were\nalso found between the tympanometrically determined peak total\ncompliance and |ECR|&lt;sup&gt;2&lt;/sup&gt; and &lt;i&gt;V&lt;sub&gt;U&lt;/sub&gt;&lt;/i&gt; at frequencies below 1 kHz. The results\nsuggest that middle ear compliance contributes significantly to the\nmeasured power reflectance and umbo velocity at frequencies below 1\nkHz but not at higher frequencies. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Conclusions: &lt;/strong&gt;  This study has established a database of objective\nmeasurements of middle ear function (ECR, umbo velocity, tympanometry)\nin a population of strictly defined normal-hearing ears. These data\nwill promote our understanding of normal middle ear function and will\nserve as a control for comparison to similar measurements made in\npathological ears. &lt;/p&gt;&quot;
## 7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      &quot;&lt;document&gt; &lt;p&gt; &lt;strong&gt; &lt;i&gt; Objective: &lt;/i&gt; &lt;/strong&gt; This study examined differences between\nthe middle ears of two ethnic groups, Caucasian\nand Chinese young adults with normal hearing,\nusing a new middle-ear measurement technique,\nwideband energy reflectance. The goal of this study\nwas to determine whether the Chinese group had\ndifferent middle-ear transmission properties than\nthe Caucasian group. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; &lt;i&gt; Design: &lt;/i&gt; &lt;/strong&gt; There were 126 subjects (237 ears) between\nthe ages of 18 and 32 yr, with 62 subjects in the\nCaucasian group and 64 subjects in the Chinese\ngroup. Wideband energy reflectance data were gathered\nusing Mimosa Acoustics (RMS system version\n4.0.4.4) wideband reflectance (WBR) equipment. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; &lt;i&gt; Results: &lt;/i&gt; &lt;/strong&gt; The Chinese group had significantly lower\nwideband energy reflectance than their Caucasian\ncounterparts at higher frequencies; however, the\nCaucasian group had significantly lower energy reflectance\nat lower frequencies than the Chinese\ngroup. The Chinese group also had significantly\nlower admittance magnitude than the Caucasian\ngroup at lower frequencies. Because body size indices\nare more comparable between Caucasian females and\nChinese males, the effect of body size could be potentially\nadjusted for by comparing the Caucasian female\nsubjects with the Chinese male subjects. The\ndifferences observed between the Caucasian and the\nChinese groups were no longer significant when the\nCaucasian female subjects were compared with the\nChinese male subjects. Applying the Caucasian norms\nto four Caucasian adults with surgically confirmed\notosclerosis resulted in an improved hit rate compared\nwith the combined Caucasian and Chinese\nnorms and the Chinese-only norms. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; &lt;i&gt; Conclusions: &lt;/i&gt; &lt;/strong&gt;  Body size may play a role in the observed\ndifferences between the Caucasian and Chinese\ngroups. The findings of this study suggest that\nfurther research is needed to investigate the effects\nof body size on wideband energy reflectance. It\nshould be noted that factors other than body size\nmay contribute to the observed differences. Chinese\nindividuals may simply have different middle-ear\ncharacteristics than Caucasian individuals, which\ncould affect WBR. In the meantime, overall test\nperformance may be improved by using a more homogeneous norm when evaluating the middle ear\nof Caucasian or Chinese individuals with WBR. &lt;/p&gt;\n&lt;p&gt;&lt;small&gt;(Ear &amp; Hearing 2006;27;774-788)&lt;/small&gt;&lt;/p&gt; &lt;/document&gt;&quot;
## 8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               &quot;&lt;p&gt; The wideband energy reflectance (ER) technique has become popular as a tool for evaluating middle\near function. Negative middle ear pressure (MEP) is a prevalent form of middle ear dysfunction,\nwhich may impact application of ER measurements in differential diagnosis. A negative MEP may\nbe countervailed by application of an equivalent negative ear canal pressure. The present study\nexamined ER in the same ears under normal and experimentally induced negative MEP conditions.\nThirty-five subjects produced at least one negative MEP each (&amp;#x2212;40 to &amp;#x2212;225 daPa). Negative MEP\nsignificantly altered ER in a frequency-specific manner that varied with MEP magnitude. ER\nincreased for low- to mid-frequencies with the largest change (~0.20 to 0.40) occurring between 1\nand 1.5 kHz. ER decreased for frequencies above 3 kHz with the largest change (~&amp;#x2212;0.10 to &amp;#x2212;0.25)\nobserved between 4.5 and 5.5 kHz. Magnitude of changes increased as MEP became more negative,\nas did the frequencies at which maximum changes occurred, and the frequency at which enhancement\ntransitioned to reduction. Ear canal pressure compensation restored ER to near baseline values.\nThis suggests that the compensation procedure adequately mitigates the effects of negative\nMEP on ER. Theoretical issues and clinical implications are discussed.&lt;/p&gt;&quot;
## 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          &quot;Purpose: The purpose of this study was to present normative data of tympanometric measurements of wideband acoustic immittance and to characterize wideband tympanograms.\n\nMethod: Data were collected in 84 young adults with strictly defined normal hearing and middle ear status. Energy absorbance (EA) was measured using clicks for 1/12-octave frequencies (0.236 to 8 kHz), with the ear canal air pressure systematically varied (+200 to -300 daPa). In 40 ears, 7 consecutive trials and a trial of clinical 226-Hz acoustic admittance (Ya) tympanometry followed. A cavity test was also conducted.\n\nResults: From the wideband EA tympanogram, several EA spectrums and EA tympanograms were derived. Descriptive statistics were performed, and population parameters were estimated. The immediate test-retest reliability was excellent. Effects of ear canal air pressure on EA were examined comprehensively. Differences in EA between tympanometric and ambient-pressure measurements were significant. Single-frequency EA tympanograms exemplified for half-octave frequencies were contrasted. The bandpass EA tympanogram, 0.236- and 1-kHz EA and Ya tympanograms, and 226-Hz Ya tympanogram were compared in 9 variables.\n\nConclusions: This study established a database of wideband tympanograms in healthy adults. The data analyses will promote our understanding of the middle ear transfer function. These data will serve as a reference for further studies in clinical populations.&quot;
## 10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               &quot;&lt;p&gt; The pressure reflectance &lt;i&gt;R (&amp;omega;)&lt;/i&gt; is the transfer function which may be defined for a linear one-port\nnetwork by the ratio of the reflected complex pressure divided by the incident complex pressure.\nThe reflectance is a function that is closely related to the impedance of the 1-port. The energy\nreflectance &lt;i&gt;&amp;#8475;(&amp;omega;)&lt;/i&gt; is defined as &lt;i&gt; |R|&lt;sup&gt;2&lt;/sup&gt;&lt;/i&gt;. It represents the ratio of reflected to incident energy. In the\nhuman ear canal the energy reflectance is important because it is a measure of the inefficiency of\nthe middle ear and cochlea, and because of the insight provided by its simple frequency domain\ninterpretation. One may characterize the ear canal impedance by use of the pressure reflectance\nand its magnitude, sidestepping the difficult problems of (a) the unknown canal length from the\nmeasurement point to the eardrum, (b) the complicated geometry of the drum, and (c) the\ncross-sectional area changes in the canal as a function of distance. Reported here are acoustic\nimpedancem easurements looking into the ear canal, measured on ten young adults with normal \nhearing (ages 18-24). The measurement point in the canal was approximately 0.85 cm from the\nentrance of the canal. From these measurements, the pressure reflectance in the canal is\ncomputed and impedance and reflectance measurements from 0.1 to 15.0 kHz are compared\namong ears. The average reflectance and the standard deviation of the reflectance for the ten\nsubjects have been determined. The impedance and reflectance of two common ear simulators,\nthe Br&amp;uuml;el &amp; Kjaer 4157 and the Industrial Research Products DB-100 (Zwislocki) coupler are\nalso measured and compared to the average human measurements. All measurements are made\nusing controls that assure a uniform accuracy in the acoustic calibration across subjects. This is\ndone by the use of two standard acoustic resistors whose impedances are known. From the\nexperimental results, it is concluded that there is significant subject variability in the magnitude\nof the reflectance for the ten ear canals. This variability is believed to be due to cochlear and\nmiddle ear impedance differences. An attempt was made at modeling the reflectance but, as\ndiscussed in the paper, several problems presently stand in the way of these models. Such models\nwould be useful for acoustic virtual-reality systems and for active noise control earphones.&lt;/p&gt;&quot;
## 11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               &quot;&lt;p&gt; Several studies have demonstrated that the auditory system is sensitive to changes in posture, presumably\nthrough changes in intracranial pressure (ICP) that in turn alter the intracochlear pressure, which\naffects the stiffness of the middle-ear system. This observation has led to efforts to develop an ear-canal\nbased noninvasive diagnostic measure for monitoring ICP, which is currently monitored invasively via\naccess through the skull or spine. Here, we demonstrate the effects of postural changes, and presumably\nICP changes, on distortion product otoacoustic emissions (DPOAE) magnitude, DPOAE angle, and power\nreflectance. Measurements were made on 12 normal-hearing subjects in two postural positions: upright\nat 90&amp;#176; and tilted at -45&amp;#176; to the horizontal. Measurements on each subject were repeated five times across\nfive separate measurement sessions. All three measures showed significant changes (&lt;i&gt;p&lt;/i&gt; &lt; 0.001) between\nupright and tilted for frequencies between 500 and 2000 Hz, and DPOAE angle changes were significant\nat all measured frequencies (500-4000 Hz). Intra-subject variability, assessed via standard deviations for\neach subject&#39;s multiple measurements, were generally smaller in the upright position relative to the\ntilted position.&lt;/p&gt;&quot;
## 12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      &quot;&lt;p&gt; &lt;strong&gt; Objectives: &lt;/strong&gt; Wideband acoustic transfer functions (WATF) measured in\nthe ear canal have been shown to be effective in the diagnosis of middle\near dysfunction in adults and in newborn infants. Although these\nmeasures would be diagnostically useful in older infants, normative data\non a large number of older infants are lacking. The goal of this study\nwas to provide such normative data. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Design: &lt;/strong&gt; The WATF of 458 infants aged 2 to 9 mos and of 210 adults\nwere obtained. Wideband reactance (X), resistance (R), and energy\nreflectance (ER) were measured in third-octave bands between 250 and\n8000 Hz. The effects of age and gender on the WATF were examined,\nand the WATF in the left and right ears were compared. Test-retest\nreliability was assessed, and the relationship between the 226-Hz\ntympanogram and the WATF was examined. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Results: &lt;/strong&gt; The results agreed well with previous reports testing fewer\nsubjects, which documented age-related change in these measures\nduring infancy and between infancy and adulthood. Test-retest correlations\nwithin third octaves were 0.5 to 0.7 at best, but did not vary\nsystematically with age. Infants&#39; test-retest absolute differences within\nthird octaves for R and ER were similar to those of adults. The shape of\nthe WATF on retest was highly repeatable, and the shapes of the WATF\nin the ears of the same individual were qualitatively similar. The\nwideband impedance results were not different in the left and right ears,\nbut ER was slightly, but significantly, lower in the left ear than that in the\nright ear. Resistance and reactance magnitude were greater for females\nthan males, but there was no effect of gender on ER. Infants whose\n226-Hz tympanogram indicated reduced peak admittance (Types As and\nB) had greater resistance and reactance magnitude than those with\nnormal peak admittance (Types A and C), but no tympanometry group\ndifferences were evident in ER. &lt;/p&gt;\n\n&lt;p&gt; &lt;strong&gt; Conclusions: &lt;/strong&gt;  Age-graded norms are essential to the successful clinical\napplication of WATF. However, the effects of gender and laterality on the\nWATF are small.&lt;/p&gt;&quot;
##                      DataSubmitterName
## 1                           Susan Voss
## 2  M. Patrick Feeney; Douglas H. Keefe
## 3                          Steve Neely
## 4                          James Lewis
## 5                        Douglas Keefe
## 6                        John Rosowski
## 7                        Navid Shahnaz
## 8                        Xiao-Ming Sun
## 9                        Xiao-Ming Sun
## 10                          Susan Voss
## 11                          Susan Voss
## 12                       Douglas Keefe
##                                   DataSubmitterEmail DateSubmitted
## 1                                    svoss@smith.edu       8/24/16
## 2  Patrick.Feeney@va.gov; Douglas.Keefe@boystown.org        6/7/18
## 3                         Stephen.Neely@boystown.org       6/18/19
## 4                                  jdlewis@uthsc.edu      10/10/18
## 5                         Douglas.Keefe@boystown.org       6/26/18
## 6                     John_Rosowski@meei.harvard.edu       11/6/15
## 7                        nshahnaz@audiospeech.ubc.ca       8/24/16
## 8                          xiao-ming.sun@wichita.edu       10/6/18
## 9                          xiao-ming.sun@wichita.edu      10/31/17
## 10                                   svoss@smith.edu        2/8/17
## 11                                   svoss@smith.edu        6/5/18
## 12                        Douglas.Keefe@boystown.org        9/1/17
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         PI_Notes
## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Measurements made on 7 subjects across multiple sessions and 3 probe locations for each subject.  Database includes measurements at only the deepest insertion depth (Position 1) and Channel B only.
## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Database includes measurements on 32 subjects, most with left and right ears and most with two sessions.  Each session includes WAI measurements at ambient ear canal, and the ear canal pressurized to TPP, negative tail pressure (-300 daPa) and positive tail pressure (200 daPa) with both &quot;Downswept&quot; and &quot;Upswept&quot; conditions.  Ear-canal areas used to calculate absorbance were estimated acoustically and are listed for the ambient and TPP measurements.  Impedance measurements at the probe location are only available for the ambient ear-canal pressure measurements and are &quot;NULL&quot; for the pressurized canal measurements.
## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Data collected on system described in Rasetshwane and Neely (2011)
## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               &quot;Custom system ER10B+ used for measurement of ear canal impedance designed by Dr. J. H. Siegel at Northwestern University.\n\nEar canal areas were estimated from the surge impedance.\n\nImpedance was measured at 2 insertion depths (deep - session 1, shallow - session 2).&quot;
## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
## 6  &quot;HearID (Mimosa Acoustics); \nNormal Criteria as follows: \n(1) There was no history of significant middle ear disease (e.g., otitis media or effusion 2 or more years previously were not considered significant if there were no known residual consequences).\n(2) There was no history of otologic surgery, with the exception of myringotomy or tympanostomy tube placement over 2 yr prior. \n(3) The external ear and TM revealed no abnormalities on otoscopic examination. \n(4) Audiometric measurements had pure-tone thresholds of 20 dB HL or better at octave frequen- cies between 0.250 and 8 kHz. \n(5) Air-bone gaps were no greater than 15 dB at 0.25 kHz and 10 dB between frequencies of 0.5 to 4 kHz. Most subjects had air and bone thresholds between 0 and 10 dB HL with an average near 8 to 9 dB HL at the highest frequencies. \n(6) Tympanograms were Type-A peaked, with peak pressures of 100 to 50 daPa, static compliance of 0.3 to 2.0 cc, total tympanometric volumes (static compliance ear canal volume) between 0.7 and 2.7 cc, and normal-appearing shape that is neither rounded nor sharp. \n(7) All subjects included in the &quot;&quot;normal hearing&quot;&quot; population were required to have two &quot;&quot;normal&quot;&quot; ears (as defined by criteria described earlier). &quot;
## 7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
## 8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         A research version of Titan (Interacoustics) was used. In this study, a total of five reflectance measurements at ambient pressure were taken per ear (detailed in the article).  Included in this database is the baseline session, defined as C in the paper, as the normative data.
## 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
## 10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
## 11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
## 12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Used an ER-1 earphone and ER-7C microphone.  Data provided by Doug Keefe and formatted by Susan Voss with help.  Lynne Werner is retired.</code></pre>
<div class="sourceCode" id="cb575"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb575-1" data-line-number="1"><span class="co"># be careful with collect() when dealing with large tables!</span></a></code></pre></div>
<p>Note how the number of rows is unknown (<code>??</code> at the top of the output above) for the lazy query.</p>
<p>Similarly, we can explore the <code>Subjects</code> table.</p>
<div class="sourceCode" id="cb576"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb576-1" data-line-number="1">Subject  <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">total =</span> <span class="kw">n</span>())</a></code></pre></div>
<pre><code>## # Source:   lazy query [?? x 1]
## # Database: mysql 5.5.58-0ubuntu0.14.04.1-log [waiuser@scidb.smith.edu:/wai]
##   total
##   &lt;dbl&gt;
## 1   640</code></pre>
<div class="sourceCode" id="cb578"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb578-1" data-line-number="1">Subject <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">collect</span>()  <span class="co"># be careful with collect() with large tables!</span></a></code></pre></div>
<pre><code>## # A tibble: 640 x 10
##    Identifier Sub_Number Session_Total   Age Female  Race Ethnicity
##    &lt;chr&gt;      &lt;chr&gt;              &lt;int&gt; &lt;int&gt;  &lt;int&gt; &lt;int&gt;     &lt;int&gt;
##  1 Voss_2010  1                      5    20      1     0         0
##  2 Voss_2010  2                      5    39      1     0         0
##  3 Voss_2010  3                      5    18      1     0         0
##  4 Voss_2010  4                      5    19      1     0         0
##  5 Voss_2010  5                      5    21      1     0         0
##  6 Voss_2010  6                      5    21      1     0         0
##  7 Voss_2010  7                      5    21      1     0         0
##  8 Voss_2010  8                      5    42      1     0         0
##  9 Voss_2010  9                      5    38      0     0         0
## 10 Voss_2010  10                     5    20      1     0         0
## # … with 630 more rows, and 3 more variables: Left_Ear_Status &lt;int&gt;,
## #   Right_Ear_Status &lt;int&gt;, Sub_Notes &lt;chr&gt;</code></pre>
<p>Let’s explore the <code>Measurements</code> table.</p>
<div class="sourceCode" id="cb580"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb580-1" data-line-number="1">Measurements <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">total =</span> <span class="kw">n</span>())</a></code></pre></div>
<pre><code>## # Source:   lazy query [?? x 1]
## # Database: mysql 5.5.58-0ubuntu0.14.04.1-log [waiuser@scidb.smith.edu:/wai]
##    total
##    &lt;dbl&gt;
## 1 286774</code></pre>
<p>There are more than a quarter million observations.</p>
<p>In the next step, we will download the data from a given subject for a specific study, in this case a paper by Rosowski et al. (2012) entitled <a href="https://www.ncbi.nlm.nih.gov/pubmed/21857517">“Ear-canal reflectance, umbo velocity, and tympanometry in normal-hearing adults”</a>.</p>
<p>Arbitrarily we choose to collect data from subject number three.</p>
<div class="sourceCode" id="cb582"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb582-1" data-line-number="1">onesubj &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb582-2" data-line-number="2"><span class="st">  </span>Measurements <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb582-3" data-line-number="3"><span class="st">  </span><span class="kw">filter</span>(Identifier <span class="op">==</span><span class="st"> &quot;Rosowski_2012&quot;</span>, Sub_Number <span class="op">==</span><span class="st"> </span><span class="dv">3</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb582-4" data-line-number="4"><span class="st">  </span>collect <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb582-5" data-line-number="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">SessionNum =</span> <span class="kw">as.factor</span>(Session))</a>
<a class="sourceLine" id="cb582-6" data-line-number="6"><span class="kw">head</span>(onesubj)</a></code></pre></div>
<pre><code>## # A tibble: 6 x 12
##   Identifier Sub_Number Session Left_Ear   MEP Instrument  Freq Absorbance
##   &lt;chr&gt;           &lt;int&gt;   &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;      &lt;int&gt; &lt;dbl&gt;      &lt;dbl&gt;
## 1 Rosowski_…          3       1        1    NA          1  211.     0.0852
## 2 Rosowski_…          3       1        1    NA          1  234.     0.0903
## 3 Rosowski_…          3       1        1    NA          1  258.     0.112 
## 4 Rosowski_…          3       1        1    NA          1  281.     0.103 
## 5 Rosowski_…          3       1        1    NA          1  305.     0.129 
## 6 Rosowski_…          3       1        1    NA          1  328.     0.136 
## # … with 4 more variables: Zmag &lt;dbl&gt;, Zang &lt;dbl&gt;, Canal_Area &lt;dbl&gt;,
## #   SessionNum &lt;fct&gt;</code></pre>
<p>Finally we can display the results of the measurements as a function of frequency and which ear (left or right) that was used.</p>
<div class="sourceCode" id="cb584"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb584-1" data-line-number="1">onesubj &lt;-<span class="st"> </span><span class="kw">mutate</span>(onesubj, <span class="dt">Ear =</span> <span class="kw">ifelse</span>(Left_Ear <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;Left&quot;</span>, <span class="st">&quot;Right&quot;</span>))</a>
<a class="sourceLine" id="cb584-2" data-line-number="2"><span class="kw">ggplot</span>(onesubj, <span class="kw">aes</span>(<span class="dt">x =</span> Freq, <span class="dt">y =</span> Absorbance)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb584-3" data-line-number="3"><span class="st">  </span><span class="kw">aes</span>(<span class="dt">colour =</span> Ear) <span class="op">+</span><span class="st"> </span><span class="kw">scale_x_log10</span>() <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;Absorbance by ear Rosowski subject 3&quot;</span>)</a></code></pre></div>
<p><img src="10-misc_files/figure-html/unnamed-chunk-46-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Always a good idea to terminate the SQL connection.</p>
<div class="sourceCode" id="cb585"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb585-1" data-line-number="1"><span class="kw">dbDisconnect</span>(con)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>We note that a number of relational database systems exist, including MySQL (illustrated here), PostgreSQL, and SQLite. More information about databases within R can be found in the CRAN <a href="https://cran.r-project.org/web/views/Databases.html">Databases with R</a> Task View.</p>
<p>Setting up and managing a database is a topic for a different day: here we focused on how SQL can be used within R to access data in a flexible and powerful manner.</p>
</div>
<div id="learn-more-2" class="section level4 unnumbered">
<h4>Learn more</h4>
<ul>
<li><a href="https://chance.amstat.org/2015/04/setting-the-stage/" class="uri">https://chance.amstat.org/2015/04/setting-the-stage/</a> (Setting the stage for data technologies)</li>
<li><a href="https://www.w3schools.com/sql/sql_intro.asp" class="uri">https://www.w3schools.com/sql/sql_intro.asp</a> (Intro to SQL)</li>
<li><a href="http://www.science.smith.edu/wai-database/home/about/" class="uri">http://www.science.smith.edu/wai-database/home/about/</a> (WAI SQL Database)</li>
<li><a href="https://cran.r-project.org/web/views/Databases.html" class="uri">https://cran.r-project.org/web/views/Databases.html</a> (CRAN Task View on Databases with R)</li>
<li><a href="https://db.rstudio.com" class="uri">https://db.rstudio.com</a> (RStudio Database resources)</li>
<li><a href="https://dbplyr.tidyverse.org/articles/dbplyr.html" class="uri">https://dbplyr.tidyverse.org/articles/dbplyr.html</a> (dbplyr package)</li>
</ul>
</div>
</div>
<div id="Dec10" class="section level2">
<h2><span class="header-section-number">10.7</span> 12/10/19 Agenda</h2>
<ol style="list-style-type: decimal">
<li>Regular Expressions</li>
<li><em>Computational</em> and <em>Statistics</em></li>
</ol>
</div>
<div id="regexpr" class="section level2">
<h2><span class="header-section-number">10.8</span> Regular Expressions</h2>
<blockquote>
<p>A regular expression … is a sequence of characters that define a search pattern. Usually such patterns are used by string searching algorithms for “find” or “find and replace” operations on strings, or for input validation. It is a technique developed in theoretical computer science and formal language theory. [From <a href="https://en.wikipedia.org/wiki/Regular_expression" class="uri">https://en.wikipedia.org/wiki/Regular_expression</a>]</p>
</blockquote>
<div id="main-tasks-in-character-matching" class="section level4 unnumbered">
<h4>Main tasks in character matching:</h4>
<ol style="list-style-type: decimal">
<li>basic string operations</li>
<li>pattern matching (regular expressions)</li>
<li>sentiment analysis</li>
</ol>
<p>The ideas below are mostly taken from Jenny Bryan’s STAT545 class: <a href="https://github.com/STAT545-UBC/STAT545-UBC-original-website/blob/master/block022_regular-expression.rmd" class="uri">https://github.com/STAT545-UBC/STAT545-UBC-original-website/blob/master/block022_regular-expression.rmd</a></p>
</div>
<div id="r-packages-to-make-your-life-easier" class="section level4 unnumbered">
<h4>R packages to make your life easier</h4>
<ul>
<li><a href="https://stringr.tidyverse.org/"><code>stringr</code> package</a> A core package in the <code>tidyverse</code>. It is installed via <code>install.packages(&quot;tidyverse&quot;)</code> and also loaded via <code>library(tidyverse)</code>. Of course, you can also install or load it individually.
<ul>
<li>Many of the main functions start with <code>str_</code>. Auto-complete is your friend.</li>
<li>Replacements for base functions re: string manipulation and regular expressions (see below).</li>
<li>Main advantages over base functions: greater consistency about inputs and outputs. Outputs are more ready for your next analytical task.</li>
<li><code>stringr</code> cheat sheet: <a href="https://github.com/rstudio/cheatsheets/raw/master/strings.pdf" class="uri">https://github.com/rstudio/cheatsheets/raw/master/strings.pdf</a></li>
</ul></li>
<li><a href="https://tidyr.tidyverse.org/"><code>tidyr</code> package</a> Especially useful for functions that split one character vector into many and vice versa: <code>separate()</code>, <code>unite()</code>, <code>extract()</code>.</li>
<li>Base functions: <code>nchar()</code>, <code>strsplit()</code>, <code>substr()</code>, <code>paste()</code>, <code>paste0()</code>.</li>
<li>The <code>glue</code> package is fantastic for string interpolation. If <code>stringr::str_interp()</code> doesn’t get your job done, check out the <code>glue</code> package.</li>
</ul>
</div>
<div id="string-functions-related-to-regular-expression" class="section level4 unnumbered">
<h4>String functions related to regular expression</h4>
<p>Regular expression is a pattern that describes a specific set of strings with a common structure. It is heavily used for string matching / replacing in all programming languages, although specific syntax may differ a bit. It is truly the heart and soul for string operations. In R, many string functions in <code>base</code> R as well as in <code>stringr</code> package use regular expressions, even Rstudio’s search and replace allows regular expression:</p>
<ul>
<li>identify match to a pattern: <code>grep(..., value = FALSE)</code>, <code>grepl()</code>, <code>stringr::str_detect()</code></li>
<li>extract match to a pattern: <code>grep(..., value = TRUE)</code>, <code>stringr::str_extract()</code>, <code>stringr::str_extract_all()</code><br />
</li>
<li>locate pattern within a string, i.e. give the start position of matched patterns. <code>regexpr()</code>, <code>gregexpr()</code>, <code>stringr::str_locate()</code>, <code>string::str_locate_all()</code><br />
</li>
<li>replace a pattern: <code>sub()</code>, <code>gsub()</code>, <code>stringr::str_replace()</code>, <code>stringr::str_replace_all()</code><br />
</li>
<li>split a string using a pattern: <code>strsplit()</code>, <code>stringr::str_split()</code></li>
</ul>
<p>Regular expressions typically specify characters (or character classes) to seek out, possibly with information about repeats and location within the string. This is accomplished with the help of metacharacters that have specific meaning: <code>$ * + . ? [ ] ^ { } | ( ) \</code>. We will use some small examples to introduce regular expression syntax and what these metacharacters mean.</p>
</div>
<div id="escape-sequences" class="section level4 unnumbered">
<h4>Escape sequences</h4>
<p>There are some special characters in R that cannot be directly coded in a string. For example, let’s say you specify your pattern with single quotes and you want to find countries with the single quote <code>'</code>. You would have to “escape” the single quote in the pattern, by preceding it with <code>\</code>, so it is clear that it is not part of the string-specifying machinery.</p>
<p>There are other characters in R that require escaping, and this rule applies to all string functions in R, including regular expressions. See <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/Quotes.html">here</a> for a complete list of R escape sequences.</p>
<ul>
<li><code>\'</code>: single quote. You don’t need to escape single quote inside a double-quoted string, so we can also use <code>&quot;'&quot;</code> in the previous example.<br />
</li>
<li><code>\&quot;</code>: double quote. Similarly, double quotes can be used inside a single-quoted string, i.e. <code>'&quot;'</code>.<br />
</li>
<li><code>\n</code>: newline.<br />
</li>
<li><code>\r</code>: carriage return.<br />
</li>
<li><code>\t</code>: tab character.</li>
</ul>
<blockquote>
<p>Note: <code>cat()</code> and <code>print()</code> handle escape sequences differently, if you want to print a string out with these sequences interpreted, use <code>cat()</code>.</p>
</blockquote>
<div class="sourceCode" id="cb587"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb587-1" data-line-number="1"><span class="kw">print</span>(<span class="st">&quot;a</span><span class="ch">\n</span><span class="st">b&quot;</span>)</a></code></pre></div>
<pre><code>## [1] &quot;a\nb&quot;</code></pre>
<div class="sourceCode" id="cb589"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb589-1" data-line-number="1"><span class="kw">cat</span>(<span class="st">&quot;a</span><span class="ch">\n</span><span class="st">b&quot;</span>)</a></code></pre></div>
<pre><code>## a
## b</code></pre>
</div>
<div id="quantifiers" class="section level4 unnumbered">
<h4>Quantifiers</h4>
<p>Quantifiers specify how many repetitions of the pattern.</p>
<ul>
<li><code>*</code>: matches at least 0 times.<br />
</li>
<li><code>+</code>: matches at least 1 times.<br />
</li>
<li><code>?</code>: matches at most 1 times.<br />
</li>
<li><code>{n}</code>: matches exactly n times.<br />
</li>
<li><code>{n,}</code>: matches at least n times.<br />
</li>
<li><code>{n,m}</code>: matches between n and m times.</li>
</ul>
<div class="sourceCode" id="cb591"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb591-1" data-line-number="1">(strings &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;ab&quot;</span>, <span class="st">&quot;acb&quot;</span>, <span class="st">&quot;accb&quot;</span>, <span class="st">&quot;acccb&quot;</span>, <span class="st">&quot;accccb&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;a&quot;      &quot;ab&quot;     &quot;acb&quot;    &quot;accb&quot;   &quot;acccb&quot;  &quot;accccb&quot;</code></pre>
<div class="sourceCode" id="cb593"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb593-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac*b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;ab&quot;     &quot;acb&quot;    &quot;accb&quot;   &quot;acccb&quot;  &quot;accccb&quot;</code></pre>
<div class="sourceCode" id="cb595"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb595-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac*b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## [1] 2 3 4 5 6</code></pre>
<div class="sourceCode" id="cb597"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb597-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac+b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;acb&quot;    &quot;accb&quot;   &quot;acccb&quot;  &quot;accccb&quot;</code></pre>
<div class="sourceCode" id="cb599"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb599-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac?b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;ab&quot;  &quot;acb&quot;</code></pre>
<div class="sourceCode" id="cb601"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb601-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac{2}b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;accb&quot;</code></pre>
<div class="sourceCode" id="cb603"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb603-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac{2}b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb605"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb605-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac{2,}b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;accb&quot;   &quot;acccb&quot;  &quot;accccb&quot;</code></pre>
<div class="sourceCode" id="cb607"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb607-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ac{2,3}b&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;accb&quot;  &quot;acccb&quot;</code></pre>
</div>
<div id="position-of-pattern-within-the-string" class="section level4 unnumbered">
<h4>Position of pattern within the string</h4>
<ul>
<li><code>^</code>: matches the start of the string.<br />
</li>
<li><code>$</code>: matches the end of the string.<br />
</li>
<li><code>\b</code>: matches the empty string at either edge of a <em>word</em>. Don’t confuse it with <code>^ $</code> which marks the edge of a <em>string</em>.<br />
</li>
<li><code>\B</code>: matches the empty string provided it is not at an edge of a word.</li>
</ul>
<div class="sourceCode" id="cb609"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb609-1" data-line-number="1">(strings &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;abcd&quot;</span>, <span class="st">&quot;cdab&quot;</span>, <span class="st">&quot;cabd&quot;</span>, <span class="st">&quot;c abd&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;abcd&quot;  &quot;cdab&quot;  &quot;cabd&quot;  &quot;c abd&quot;</code></pre>
<div class="sourceCode" id="cb611"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb611-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ab&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abcd&quot;  &quot;cdab&quot;  &quot;cabd&quot;  &quot;c abd&quot;</code></pre>
<div class="sourceCode" id="cb613"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb613-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;^ab&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abcd&quot;</code></pre>
<div class="sourceCode" id="cb615"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb615-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ab$&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;cdab&quot;</code></pre>
<div class="sourceCode" id="cb617"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb617-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">bab&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abcd&quot;  &quot;c abd&quot;</code></pre>
</div>
<div id="operators" class="section level4 unnumbered">
<h4>Operators</h4>
<ul>
<li><code>.</code>: matches any single character, as shown in the first example.</li>
<li><code>[...]</code>: a character list, matches any one of the characters inside the square brackets. We can also use <code>-</code> inside the brackets to specify a range of characters.<br />
</li>
<li><code>[^...]</code>: an inverted character list, similar to <code>[...]</code>, but matches any characters <strong>except</strong> those inside the square brackets.<br />
</li>
<li><code>\</code>: suppress the special meaning of metacharacters in regular expression, i.e. <code>$ * + . ? [ ] ^ { } | ( ) \</code>, similar to its usage in escape sequences. Since <code>\</code> itself needs to be escaped in R, we need to escape these metacharacters with double backslash like <code>\\$</code>.<br />
</li>
<li><code>|</code>: an “or” operator, matches patterns on either side of the <code>|</code>.<br />
</li>
<li><code>(...)</code>: grouping in regular expressions. This allows you to retrieve the bits that matched various parts of your regular expression so you can alter them or use them for building up a new string. Each group can than be refer using <code>\\N</code>, with N being the No. of <code>(...)</code> used. This is called <strong>backreference</strong>.</li>
</ul>
<div class="sourceCode" id="cb619"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb619-1" data-line-number="1">(strings &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;^ab&quot;</span>, <span class="st">&quot;ab&quot;</span>, <span class="st">&quot;abc&quot;</span>, <span class="st">&quot;abd&quot;</span>, <span class="st">&quot;abe&quot;</span>, <span class="st">&quot;ab 12&quot;</span>))</a></code></pre></div>
<pre><code>## [1] &quot;^ab&quot;   &quot;ab&quot;    &quot;abc&quot;   &quot;abd&quot;   &quot;abe&quot;   &quot;ab 12&quot;</code></pre>
<div class="sourceCode" id="cb621"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb621-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ab.&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abc&quot;   &quot;abd&quot;   &quot;abe&quot;   &quot;ab 12&quot;</code></pre>
<div class="sourceCode" id="cb623"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb623-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ab[c-e]&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abc&quot; &quot;abd&quot; &quot;abe&quot;</code></pre>
<div class="sourceCode" id="cb625"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb625-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;ab[^c]&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abd&quot;   &quot;abe&quot;   &quot;ab 12&quot;</code></pre>
<div class="sourceCode" id="cb627"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb627-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;^ab&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;ab&quot;    &quot;abc&quot;   &quot;abd&quot;   &quot;abe&quot;   &quot;ab 12&quot;</code></pre>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb629-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">^ab&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;^ab&quot;</code></pre>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb631-1" data-line-number="1"><span class="kw">grep</span>(<span class="st">&quot;abc|abd&quot;</span>, strings, <span class="dt">value =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## [1] &quot;abc&quot; &quot;abd&quot;</code></pre>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb633-1" data-line-number="1"><span class="kw">gsub</span>(<span class="st">&quot;(ab) 12&quot;</span>, <span class="st">&quot;</span><span class="ch">\\</span><span class="st">1 34&quot;</span>, strings)</a></code></pre></div>
<pre><code>## [1] &quot;^ab&quot;   &quot;ab&quot;    &quot;abc&quot;   &quot;abd&quot;   &quot;abe&quot;   &quot;ab 34&quot;</code></pre>
</div>
<div id="character-classes" class="section level4 unnumbered">
<h4>Character classes</h4>
<p>Character classes allow specifying entire classes of characters, such as numbers, letters, etc. There are two flavors of character classes, one uses <code>[:</code> and <code>:]</code> around a predefined name inside square brackets and the other uses <code>\</code> and a special character. They are sometimes interchangeable.</p>
<ul>
<li><code>[:digit:]</code> or <code>\d</code>: digits, 0 1 2 3 4 5 6 7 8 9, equivalent to <code>[0-9]</code>.<br />
</li>
<li><code>\D</code>: non-digits, equivalent to <code>[^0-9]</code>.<br />
</li>
<li><code>[:lower:]</code>: lower-case letters, equivalent to <code>[a-z]</code>.<br />
</li>
<li><code>[:upper:]</code>: upper-case letters, equivalent to <code>[A-Z]</code>.<br />
</li>
<li><code>[:alpha:]</code>: alphabetic characters, equivalent to <code>[[:lower:][:upper:]]</code> or <code>[A-z]</code>.<br />
</li>
<li><code>[:alnum:]</code>: alphanumeric characters, equivalent to <code>[[:alpha:][:digit:]]</code> or <code>[A-z0-9]</code>.<br />
</li>
<li><code>\w</code>: word characters, equivalent to <code>[[:alnum:]_]</code> or <code>[A-z0-9_]</code>.<br />
</li>
<li><code>\W</code>: not word, equivalent to <code>[^A-z0-9_]</code>.<br />
</li>
<li><code>[:xdigit:]</code>: hexadecimal digits (base 16), 0 1 2 3 4 5 6 7 8 9 A B C D E F a b c d e f, equivalent to <code>[0-9A-Fa-f]</code>.</li>
<li><code>[:blank:]</code>: blank characters, i.e. space and tab.<br />
</li>
<li><code>[:space:]</code>: space characters: tab, newline, vertical tab, form feed, carriage return, space.</li>
<li><code>\s</code>: space, <code></code>.<br />
</li>
<li><code>\S</code>: not space.<br />
</li>
<li><code>[:punct:]</code>: punctuation characters, ! &quot; # $ % &amp; ’ ( ) * + , - . / : ; &lt; = &gt; ? @ [  ] ^ _ ` { | } ~.</li>
<li><code>[:graph:]</code>: graphical (human readable) characters: equivalent to <code>[[:alnum:][:punct:]]</code>.</li>
<li><code>[:print:]</code>: printable characters, equivalent to <code>[[:alnum:][:punct:]\\s]</code>.</li>
<li><code>[:cntrl:]</code>: control characters, like <code>\n</code> or <code>\r</code>, <code>[\x00-\x1F\x7F]</code>.</li>
</ul>
<p>Note:<br />
* <code>[:...:]</code> has to be used inside square brackets, e.g. <code>[[:digit:]]</code>.<br />
* <code>\</code> itself is a special character that needs escape, e.g. <code>\\d</code>. Do not confuse these regular expressions with R escape sequences such as <code>\t</code>.</p>
</div>
<div id="stringr" class="section level4 unnumbered">
<h4><code>stringr</code></h4>
<p>In many cases, you will want to use the incredibly useful and tidy set of functions available in the <code>stringr</code> package. (<code>stringr</code> is a core package in the <code>tidyverse</code>.) For example, below we’ve extracted the first (and then last) word as a character string from the <code>StreetName</code> variable.</p>
<ul>
<li><code>stringr</code> cheat sheet: <a href="https://github.com/rstudio/cheatsheets/raw/master/strings.pdf" class="uri">https://github.com/rstudio/cheatsheets/raw/master/strings.pdf</a></li>
</ul>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb635-1" data-line-number="1"><span class="kw">library</span>(Stat2Data)</a>
<a class="sourceLine" id="cb635-2" data-line-number="2"><span class="kw">data</span>(RailsTrails)</a>
<a class="sourceLine" id="cb635-3" data-line-number="3">RailsTrails &lt;-<span class="st"> </span>RailsTrails <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb635-4" data-line-number="4"><span class="st">  </span><span class="kw">select</span>(HouseNum, Bedrooms, Price2014, StreetName) </a>
<a class="sourceLine" id="cb635-5" data-line-number="5">RailsTrails <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>##   HouseNum Bedrooms Price2014      StreetName
## 1        1        3   210.729 Acrebrook Drive
## 2        2        3   204.171       Autumn Dr
## 3        3        3   338.662     Bridge Road
## 4        4        3   276.250     Bridge Road
## 5        5        4   169.173     Bridge Road
## 6        6        3   211.487 Brierwood Drive</code></pre>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb637-1" data-line-number="1">RailsTrails <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb637-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">first_piece =</span> stringr<span class="op">::</span><span class="kw">word</span>(StreetName, <span class="dt">start =</span> <span class="dv">1</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>##   HouseNum Bedrooms Price2014      StreetName first_piece
## 1        1        3   210.729 Acrebrook Drive   Acrebrook
## 2        2        3   204.171       Autumn Dr      Autumn
## 3        3        3   338.662     Bridge Road      Bridge
## 4        4        3   276.250     Bridge Road      Bridge
## 5        5        4   169.173     Bridge Road      Bridge
## 6        6        3   211.487 Brierwood Drive   Brierwood</code></pre>
<div class="sourceCode" id="cb639"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb639-1" data-line-number="1">RailsTrails <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb639-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">last_piece =</span> stringr<span class="op">::</span><span class="kw">word</span>(StreetName, <span class="dt">start =</span> <span class="dv">-1</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>##   HouseNum Bedrooms Price2014      StreetName last_piece
## 1        1        3   210.729 Acrebrook Drive      Drive
## 2        2        3   204.171       Autumn Dr         Dr
## 3        3        3   338.662     Bridge Road       Road
## 4        4        3   276.250     Bridge Road       Road
## 5        5        4   169.173     Bridge Road       Road
## 6        6        3   211.487 Brierwood Drive      Drive</code></pre>
</div>
<div id="an-example-from-my-work" class="section level4 unnumbered">
<h4>An example from my work</h4>
<p>Below are a handful of string characters that represent genomic sequences which were measured in an RNA Sequencing dataset. The task below is to find intergenic regions (IGR) and identify which coding sequences (CDS) bookend the intergenic regions. Note that IGRs do not code for proteins while CDSs do. Additionally, AS refers to anti-sense which identifies the genomic sequence in the opposite orientation (e.g., CGGATCC vs CCTAGGC). [The code below was written by Madison Hobbs, Scripps ’19.]</p>
<div id="the-names-of-the-genomic-pieces" class="section level5 unnumbered">
<h5>The names of the genomic pieces</h5>
<div class="sourceCode" id="cb641"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb641-1" data-line-number="1">allCounts &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Geneid =</span> <span class="kw">c</span>(<span class="st">&quot;CDS:b2743:pcm:L-isoaspartate_protein_carboxylmethyltransferase_type_II:cds2705:-:626:NC_000913.3&quot;</span>,</a>
<a class="sourceLine" id="cb641-2" data-line-number="2">            <span class="st">&quot;CDS:b2764:cysJ:sulfite_reductase2C_alpha_subunit2C_flavoprotein:cds2726:-:1799:NC_000913.3&quot;</span>,</a>
<a class="sourceLine" id="cb641-3" data-line-number="3">            <span class="st">&quot;IGR:(CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893):+:945:NC_000913.3&quot;</span>,</a>
<a class="sourceLine" id="cb641-4" data-line-number="4">            <span class="st">&quot;AS_IGR:(CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587):+:639:NC_000913.3&quot;</span>,</a>
<a class="sourceLine" id="cb641-5" data-line-number="5">            <span class="st">&quot;IGR:(CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344):+:396:NC_000913.3&quot;</span>))</a>
<a class="sourceLine" id="cb641-6" data-line-number="6"></a>
<a class="sourceLine" id="cb641-7" data-line-number="7">allCounts<span class="op">$</span>GeneidBackup =<span class="st"> </span>allCounts<span class="op">$</span>Geneid</a></code></pre></div>
<p>First, it is important to identify which are IGR, CDS, and anti-sense.</p>
<div class="sourceCode" id="cb642"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb642-1" data-line-number="1">allCounts &lt;-<span class="st"> </span>allCounts <span class="op">%&gt;%</span><span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(Geneid, <span class="kw">c</span>(<span class="st">&quot;feature&quot;</span>, <span class="st">&quot;rest&quot;</span>), <span class="dt">sep=</span><span class="st">&quot;[:]&quot;</span>)</a>
<a class="sourceLine" id="cb642-2" data-line-number="2">allCounts</a></code></pre></div>
<pre><code>##   feature
## 1     CDS
## 2     CDS
## 3     IGR
## 4  AS_IGR
## 5     IGR
##                                                                                                                                                                                        rest
## 1                                                                                                                                                                                     b2743
## 2                                                                                                                                                                                     b2764
## 3 (CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893)
## 4                                                         (CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587)
## 5                                       (CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344)
##                                                                                                                                                                                                      GeneidBackup
## 1                                                                                                                CDS:b2743:pcm:L-isoaspartate_protein_carboxylmethyltransferase_type_II:cds2705:-:626:NC_000913.3
## 2                                                                                                                      CDS:b2764:cysJ:sulfite_reductase2C_alpha_subunit2C_flavoprotein:cds2726:-:1799:NC_000913.3
## 3 IGR:(CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893):+:945:NC_000913.3
## 4                                                      AS_IGR:(CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587):+:639:NC_000913.3
## 5                                       IGR:(CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344):+:396:NC_000913.3</code></pre>
<p>We keep only the IGR and AS_IGR strings, and we separate the two bookends. Note, the separation comes at the backslash.</p>
<div class="sourceCode" id="cb644"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb644-1" data-line-number="1">igr &lt;-<span class="st"> </span>allCounts <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(feature <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;IGR&quot;</span>, <span class="st">&quot;AS_IGR&quot;</span>))</a>
<a class="sourceLine" id="cb644-2" data-line-number="2">igr &lt;-<span class="st"> </span>igr <span class="op">%&gt;%</span><span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(GeneidBackup, <span class="kw">c</span>(<span class="st">&quot;Geneid1&quot;</span>, <span class="st">&quot;Geneid2&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[/]&quot;</span>)</a>
<a class="sourceLine" id="cb644-3" data-line-number="3"><span class="kw">names</span>(igr)</a></code></pre></div>
<pre><code>## [1] &quot;feature&quot; &quot;rest&quot;    &quot;Geneid1&quot; &quot;Geneid2&quot;</code></pre>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb646-1" data-line-number="1">igr</a></code></pre></div>
<pre><code>##   feature
## 1     IGR
## 2  AS_IGR
## 3     IGR
##                                                                                                                                                                                        rest
## 1 (CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893)
## 2                                                         (CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587)
## 3                                       (CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344)
##                                                                                                            Geneid1
## 1 IGR:(CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220
## 2                                                                AS_IGR:(CDS,b0008,talB,transaldolase_B,cds7,+,953
## 3                                 IGR:(CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910
##                                                                                                    Geneid2
## 1           CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893):+:945:NC_000913.3
## 2 CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587):+:639:NC_000913.3
## 3                 CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344):+:396:NC_000913.3</code></pre>
<p>For each of the two bookend Genes, we need to separate out the feature from the rest. Note that we write over feature1 in the second line of code below. Both of the bookends for all sequences are CDS elements.</p>
<div class="sourceCode" id="cb648"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb648-1" data-line-number="1">igr<span class="op">$</span>feature1 &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(igr, Geneid1, <span class="kw">c</span>(<span class="st">&quot;feature1&quot;</span>, <span class="st">&quot;rest&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[,]&quot;</span>)<span class="op">$</span>feature1</a>
<a class="sourceLine" id="cb648-2" data-line-number="2">igr<span class="op">$</span>feature1 &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(igr, feature1, <span class="kw">c</span>(<span class="st">&quot;rest&quot;</span>, <span class="st">&quot;feature1&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[()]&quot;</span>)<span class="op">$</span>feature1</a>
<a class="sourceLine" id="cb648-3" data-line-number="3">igr<span class="op">$</span>feature2 &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(igr, Geneid2, <span class="kw">c</span>(<span class="st">&quot;feature2&quot;</span>, <span class="st">&quot;rest&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[,]&quot;</span>)<span class="op">$</span>feature2</a>
<a class="sourceLine" id="cb648-4" data-line-number="4"><span class="kw">names</span>(igr)</a></code></pre></div>
<pre><code>## [1] &quot;feature&quot;  &quot;rest&quot;     &quot;Geneid1&quot;  &quot;Geneid2&quot;  &quot;feature1&quot; &quot;feature2&quot;</code></pre>
<div class="sourceCode" id="cb650"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb650-1" data-line-number="1">igr</a></code></pre></div>
<pre><code>##   feature
## 1     IGR
## 2  AS_IGR
## 3     IGR
##                                                                                                                                                                                        rest
## 1 (CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893)
## 2                                                         (CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587)
## 3                                       (CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344)
##                                                                                                            Geneid1
## 1 IGR:(CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220
## 2                                                                AS_IGR:(CDS,b0008,talB,transaldolase_B,cds7,+,953
## 3                                 IGR:(CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910
##                                                                                                    Geneid2
## 1           CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893):+:945:NC_000913.3
## 2 CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587):+:639:NC_000913.3
## 3                 CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344):+:396:NC_000913.3
##   feature1 feature2
## 1      CDS      CDS
## 2      CDS      CDS
## 3      CDS      CDS</code></pre>
<p>As CDS, it is now important to find the actual genenames for each of the IGR sequences. We also keep each element’s <code>bnum</code> which represents a unique gene identifier in <em>E. coli</em>.</p>
<p><code>bnum</code>, <code>genename</code>, <code>rna.name</code> act as place holders for the types of elements that we will need to identify the bookends of the IGRs.</p>
<div class="sourceCode" id="cb652"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb652-1" data-line-number="1">bnum =<span class="st"> &quot;b[0-9]{4}&quot;</span></a>
<a class="sourceLine" id="cb652-2" data-line-number="2">bnum</a></code></pre></div>
<pre><code>## [1] &quot;b[0-9]{4}&quot;</code></pre>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb654-1" data-line-number="1">genename =<span class="st"> &quot;,[a-z]{3}[A-Z,].&quot;</span></a>
<a class="sourceLine" id="cb654-2" data-line-number="2">rna.name =<span class="st"> &quot;,rna[0-9]..&quot;</span></a></code></pre></div>
<div class="sourceCode" id="cb655"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb655-1" data-line-number="1">igr<span class="op">$</span>start.gene &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">case_when</span>(</a>
<a class="sourceLine" id="cb655-2" data-line-number="2">  igr<span class="op">$</span>feature1 <span class="op">==</span><span class="st"> &quot;CDS&quot;</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid1, genename),</a>
<a class="sourceLine" id="cb655-3" data-line-number="3">  <span class="ot">TRUE</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid1, rna.name))</a>
<a class="sourceLine" id="cb655-4" data-line-number="4">igr<span class="op">$</span>end.gene &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">case_when</span>(</a>
<a class="sourceLine" id="cb655-5" data-line-number="5">  igr<span class="op">$</span>feature2 <span class="op">==</span><span class="st"> &quot;CDS&quot;</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid2, genename),</a>
<a class="sourceLine" id="cb655-6" data-line-number="6">  <span class="ot">TRUE</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid2, rna.name))</a>
<a class="sourceLine" id="cb655-7" data-line-number="7">igr<span class="op">$</span>start.bnum &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">case_when</span>(</a>
<a class="sourceLine" id="cb655-8" data-line-number="8">  igr<span class="op">$</span>feature1 <span class="op">==</span><span class="st"> &quot;CDS&quot;</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid1, bnum),</a>
<a class="sourceLine" id="cb655-9" data-line-number="9">  <span class="ot">TRUE</span> <span class="op">~</span><span class="st"> &quot;none&quot;</span>)</a>
<a class="sourceLine" id="cb655-10" data-line-number="10">igr<span class="op">$</span>end.bnum &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">case_when</span>(</a>
<a class="sourceLine" id="cb655-11" data-line-number="11">  igr<span class="op">$</span>feature2 <span class="op">==</span><span class="st"> &quot;CDS&quot;</span> <span class="op">~</span><span class="st"> </span>stringr<span class="op">::</span><span class="kw">str_extract</span>(igr<span class="op">$</span>Geneid2, bnum),</a>
<a class="sourceLine" id="cb655-12" data-line-number="12">  <span class="ot">TRUE</span> <span class="op">~</span><span class="st"> &quot;none&quot;</span>)</a>
<a class="sourceLine" id="cb655-13" data-line-number="13">igr &lt;-<span class="st"> </span>igr <span class="op">%&gt;%</span><span class="st"> </span>tidyr<span class="op">::</span><span class="kw">separate</span>(start.gene, <span class="dt">into =</span> <span class="kw">c</span>(<span class="st">&quot;comma&quot;</span>, <span class="st">&quot;start.gene&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[,]&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb655-14" data-line-number="14"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>comma) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb655-15" data-line-number="15"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">separate</span>(end.gene, <span class="dt">into =</span> <span class="kw">c</span>(<span class="st">&quot;comma&quot;</span>, <span class="st">&quot;end.gene&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;[,]&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb655-16" data-line-number="16"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>comma)</a>
<a class="sourceLine" id="cb655-17" data-line-number="17"><span class="kw">names</span>(igr)</a></code></pre></div>
<pre><code>##  [1] &quot;feature&quot;    &quot;rest&quot;       &quot;Geneid1&quot;    &quot;Geneid2&quot;    &quot;feature1&quot;  
##  [6] &quot;feature2&quot;   &quot;start.gene&quot; &quot;end.gene&quot;   &quot;start.bnum&quot; &quot;end.bnum&quot;</code></pre>
<div class="sourceCode" id="cb657"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb657-1" data-line-number="1">igr</a></code></pre></div>
<pre><code>##   feature
## 1     IGR
## 2  AS_IGR
## 3     IGR
##                                                                                                                                                                                        rest
## 1 (CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220/CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893)
## 2                                                         (CDS,b0008,talB,transaldolase_B,cds7,+,953/CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587)
## 3                                       (CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910/CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344)
##                                                                                                            Geneid1
## 1 IGR:(CDS,b1594,mlc,glucosamine_anaerobic_growth_regulon_transcriptional_repressor3B_autorepressor,cds1581,-,1220
## 2                                                                AS_IGR:(CDS,b0008,talB,transaldolase_B,cds7,+,953
## 3                                 IGR:(CDS,b1808,yoaA,putative_ATP-dependent_helicase2C_DinG_family,cds1798,-,1910
##                                                                                                    Geneid2
## 1           CDS,b1595,ynfL,LysR_family_putative_transcriptional_regulator,cds1582,-,893):+:945:NC_000913.3
## 2 CDS,b0009,mog,molybdochelatase_incorporating_molybdenum_into_molybdopterin,cds8,+,587):+:639:NC_000913.3
## 3                 CDS,b1809,yoaB,putative_reactive_intermediate_deaminase,cds1799,+,344):+:396:NC_000913.3
##   feature1 feature2 start.gene end.gene start.bnum end.bnum
## 1      CDS      CDS        mlc     ynfL      b1594    b1595
## 2      CDS      CDS       talB      mog      b0008    b0009
## 3      CDS      CDS       yoaA     yoaB      b1808    b1809</code></pre>
</div>
</div>
<div id="helpful-tutorialsfiles" class="section level4 unnumbered">
<h4>Helpful tutorials/files</h4>
<ul>
<li><code>stringr</code> vignette: <a href="https://cran.r-project.org/web/packages/stringr/vignettes/stringr.html" class="uri">https://cran.r-project.org/web/packages/stringr/vignettes/stringr.html</a></li>
<li><a href="https://stringr.tidyverse.org/"><code>stringr</code> package</a></li>
<li>Jenny Bryan’s STAT 545 notes: <a href="https://stat545.com/character-vectors.html" class="uri">https://stat545.com/character-vectors.html</a></li>
<li>Jenny Bryan’s STAT 545 lab: <a href="http://stat545.com/block022_regular-expression.html" class="uri">http://stat545.com/block022_regular-expression.html</a></li>
<li>Hadley Wickham’s book <a href="https://r4ds.had.co.nz/strings.html">R for Data Science</a></li>
<li><a href="http://regexpal.com/">regexpal</a><br />
</li>
<li><a href="http://www.regexr.com/">RegExr</a><br />
</li>
<li>Regular expression in R <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/regex.html">official document</a>.</li>
</ul>
</div>
<div id="fun-examples" class="section level4 unnumbered">
<h4>Fun examples</h4>
<ul>
<li>The name Hilary: <a href="https://hilaryparker.com/2013/01/30/hilary-the-most-poisoned-baby-name-in-us-history/" class="uri">https://hilaryparker.com/2013/01/30/hilary-the-most-poisoned-baby-name-in-us-history/</a></li>
<li>Trump’s tweets: <a href="http://varianceexplained.org/r/trump-tweets/" class="uri">http://varianceexplained.org/r/trump-tweets/</a></li>
<li>Trump’s tweets, take two: <a href="http://varianceexplained.org/r/trump-followup/" class="uri">http://varianceexplained.org/r/trump-followup/</a></li>
</ul>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="unsup.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="compstat.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/10-misc.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Math-154-Notes.pdf", "Math-154-Notes.epub"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
